<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Embedding and Clustering with Alignment for Spatial Datasets • PRECAST</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="Embedding and Clustering with Alignment for Spatial Datasets">
<meta property="og:description" content="An efficient data integration method is provided for multiple spatial transcriptomics data with non-cluster-relevant effects such as the complex batch effects. It unifies spatial factor analysis simultaneously with spatial clustering and embedding alignment, requiring only partially shared cell/domain clusters across datasets. More details can be referred to Wei Liu, et al. (2023) &lt;doi:10.1038/s41467-023-35947-w&gt;.">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">PRECAST</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.6.4</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="articles/PRECAST.html">Get started</a>
</li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/PRECAST.BreastCancer.html">PRECAST: Human Breast Cancer Data Analysis</a>
    </li>
    <li>
      <a href="articles/PRECAST.DLPFC.html">PRECAST: DLPFC Single Sample Analysis</a>
    </li>
    <li>
      <a href="articles/PRECAST.DLPFC4.html">PRECAST: Four DLPFC Sample Analysis</a>
    </li>
    <li>
      <a href="articles/PRECAST.Simu.html">PRECAST: simulation</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/feiyoung/PRECAST/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">

<div class="section level1">
<div class="page-header"><h1 id="precast">PRECAST<a class="anchor" aria-label="anchor" href="#precast"></a>
</h1></div>
<p>========================================================================= <!-- badges: start --></p>
<p><a href="https://cran.r-project.org/package=PRECAST" class="external-link"><img src="https://www.r-pkg.org/badges/version-ago/PRECAST"></a> <a href="https://cran.r-project.org/package=PRECAST" class="external-link"><img src="https://cranlogs.r-pkg.org/badges/PRECAST?color=orange"></a> <a href="https://cran.r-project.org/package=PRECAST" class="external-link"><img src="https://cranlogs.r-pkg.org/badges/grand-total/PRECAST?color=orange"></a> <a href="https://zenodo.org/badge/latestdoi/500674213" class="external-link"><img src="https://zenodo.org/badge/500674213.svg" alt="DOI"></a> <!-- badges: end --></p>
<p>PRECAST: a probabilistic embedding and clustering with alignment for spatial transcriptomics data integration.</p>
<p>PRECAST is a package for integrating and analyzing multiple spatially resolved transcriptomics (SRT) datasets, developed by the Jin Liu’s lab. It unifies spatial factor analysis simultaneously with spatial clustering and embedding alignment, requiring only partially shared cell/domain clusters across datasets.</p>
<p>Check out our <a href="https://www.nature.com/articles/s41467-023-35947-w" class="external-link">Nature Communications paper</a> and our <a href="https://feiyoung.github.io/PRECAST/index.html" class="external-link">Package Website</a> for a more complete description of the methods and analyses.</p>
<p>PRECAST can be used to compare and contrast experimental datasets in a variety of contexts, for instance:</p>
<ul>
<li>Across experimental batches</li>
<li>Across individuals</li>
<li>Across different conditions (i.e., case and control)</li>
<li>Across datasets with only partially shared cell/domain clusters</li>
</ul>
<p>Once multiple datasets are integrated, the package provides functionality for further data exploration, analysis, and visualization. Users can:</p>
<ul>
<li>Identify clusters using all data information</li>
<li>Extract aligned low-dimensional embeddings across datasets</li>
<li>Recover comparable gene expression matrices among datasets</li>
<li>Find significant shared (and dataset-specific) gene markers</li>
<li>Conditional spatially variational genes analysis</li>
<li>Compare clusters with previously identified domain/cell types</li>
<li>Visuzlize extracted embeddings using 3-dim tSNE and UMAP</li>
<li>Visualize clusters and gene expression using tSNE and UMAP</li>
</ul>
</div>
<div class="section level1">
<h1 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h1>
<p>“PRECAST” depends on the ‘Rcpp’ and ‘RcppArmadillo’ package, which requires appropriate setup of computer. For the users that have set up system properly for compiling C++ files, the following installation command will work.</p>
<pre class="{rmd}"><code><span class="co"># Method 1: install PRECAST from CRAN</span>
<span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">'PRECAST'</span><span class="op">)</span>


<span class="co"># For the newest version of PRECAST, users can use method 2 for installation.</span>
<span class="co"># Method 2: Install PRECAST from Github</span>
<span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st"><a href="https://remotes.r-lib.org" class="external-link">"remotes"</a></span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span>
    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"remotes"</span><span class="op">)</span>
<span class="fu">remotes</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"feiyoung/PRECAST"</span><span class="op">)</span>

<span class="co"># If some dependent packages (such as `scater`) on Bioconductor can not be installed nomrally, use following commands, then run abouve command.</span>
<span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st"><a href="https://bioconductor.github.io/BiocManager/" class="external-link">"BiocManager"</a></span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="co">## install BiocManager</span>
    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span>
<span class="co"># install the package on Bioconducter</span>
<span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"scater"</span><span class="op">)</span><span class="op">)</span></code></pre>
<div class="section level2">
<h2 id="usage">Usage<a class="anchor" aria-label="anchor" href="#usage"></a>
</h2>
<p>For usage examples and guided walkthroughs, check the <code>vignettes</code> directory of the repo.</p>
<ul>
<li><a href="https://feiyoung.github.io/PRECAST/articles/PRECAST.DLPFC.html" class="external-link">Single SRT data analysis</a></li>
<li><a href="https://feiyoung.github.io/PRECAST/articles/PRECAST.Simu.html" class="external-link">Toy examples for integrating three batches</a></li>
<li><a href="https://feiyoung.github.io/PRECAST/articles/PRECAST.DLPFC4.html" class="external-link">Integration across experimental batches: DLPFC</a></li>
<li><a href="https://feiyoung.github.io/PRECAST/articles/PRECAST.BreastCancer.html" class="external-link">Integration across experimental batches: breast cancer</a></li>
</ul>
<p>For the users that don’t have set up system properly, the following setup on different systems can be referred. ## Setup on Windows system First, download <a href="https://cran.r-project.org/bin/windows/Rtools/" class="external-link">Rtools</a>; second, add the Rtools directory to the environment variable.</p>
</div>
<div class="section level2">
<h2 id="setup-on-macos-system">Setup on MacOS system<a class="anchor" aria-label="anchor" href="#setup-on-macos-system"></a>
</h2>
<p>First, install Xcode. Installation about Xcode can be referred <a href="https://stackoverflow.com/questions/8291146/xcode-installation-on-mac" class="external-link">here</a>.</p>
<p>Second, install “gfortran” for compiling C++ and Fortran at <a href="https://github.com/fxcoudert/gfortran-for-macOS" class="external-link">here</a>.</p>
</div>
<div class="section level2">
<h2 id="setup-on-linux-system">Setup on Linux system<a class="anchor" aria-label="anchor" href="#setup-on-linux-system"></a>
</h2>
<p>For parallel computation on Linux, users must use the following system command to set the C_stack unlimited in case of the error <code>R Error: C stack usage is too close to the limit</code>.</p>
<pre class="{linux}"><code>ulimit -s unlimited</code></pre>
<p>If you use conda environment on Linux system and some dependent packages (such as <code>scater</code>) can not normally installed, you can search R package at anaconda.org website. We take the <code>scater</code> package as example, and its search result is <a href="https://anaconda.org/bioconda/bioconductor-scater" class="external-link uri">https://anaconda.org/bioconda/bioconductor-scater</a>. Then you can install it in conda environment by following command.</p>
<pre class="{linux}"><code>
conda install -c bioconda bioconductor-scater</code></pre>
<p>For the user not using conda environment, if dependent packages (such as <code>scater</code>) not normally installed are in Bioconductor, then use the following command to install the dependent packages.</p>
<pre class="{linux}"><code><span class="co"># install BiocManager</span>
<span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st"><a href="https://bioconductor.github.io/BiocManager/" class="external-link">"BiocManager"</a></span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span>
    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span>
<span class="co"># install the package on Bioconducter</span>
<span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"scater"</span><span class="op">)</span><span class="op">)</span></code></pre>
<p>If dependent packages (such as <code>DR.SC</code>) not normally installed are in CRAN, then use the following command to install the dependent packages.</p>
<pre class="{linux}"><code><span class="co"># install the package on CRAN</span>
<span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"DR.SC"</span><span class="op">)</span></code></pre>
</div>
</div>
<div class="section level1">
<h1 id="demonstration">Demonstration<a class="anchor" aria-label="anchor" href="#demonstration"></a>
</h1>
<p>For an example of typical PRECAST usage, please see our <a href="https://feiyoung.github.io/PRECAST/index.html" class="external-link">Package Website</a> for a demonstration and overview of the functions included in PRECAST.</p>
</div>
<div class="section level1">
<h1 id="news">NEWs<a class="anchor" aria-label="anchor" href="#news"></a>
</h1>
<p>PRECAST version 1.6.4 (2024-01-24) * Update <code><a href="reference/IntegrateSpaData.html">IntegrateSpaData()</a></code> to make it able to remove unwanted variations for all genes rather than only variable genes.</p>
<p>PRECAST version 1.6.2 (2023-08-02) * Update the code to ensure compatibility with Seurat V5.</p>
<p>PRECAST version 1.6.1 (2023-05-06) * Fix the bug, reported by <code>agelber-ucsd</code>, in <code><a href="reference/IntegrateSpaData.html">IntegrateSpaData()</a></code> when adusting additional covariates.</p>
<ul><li>Note: the package in CRAN is not updated.</li></ul>
<p>PRECAST version 1.6 (2023-04-18)</p>
<ul>
<li><p>Fix the bug in <code>PRECAST:::filter_gene()</code>.</p></li>
<li><p>Revise the function name <code>selectModel()</code> to <code><a href="reference/selectModel.html">SelectModel()</a></code>, to avoid the mask when loading DR.SC package.</p></li>
<li><p>Update the tutorials.</p></li>
</ul>
<p>PRECAST version 1.5 (2023-03-05)</p>
<ul>
<li><p>Fix the <a href="https://github.com/feiyoung/PRECAST/issues/2" class="external-link">issue</a> reported by anvaly. Specifically, the assay name “RNA” used in functions <code><a href="reference/IntegrateSpaData.html">IntegrateSpaData()</a></code> is replaced by the default assay using <code>DefaultAssay</code> function in Seurat. Fix the typo <code>human = {intersect((genelist),Mouse_HK_genes$Gene)}</code> with replacement of <code>human = {intersect(toupper(genelist), PRECAST::Human_HK_genes$Gene)}</code>。</p></li>
<li><p>Add two functions for visualization: <code><a href="reference/chooseColors.html">chooseColors()</a></code> and <code><a href="reference/drawFigs.html">drawFigs()</a></code>.</p></li>
</ul>
<p>PRECAST version 1.3 (2022-10-05)</p>
<ul>
<li><p>Fix the <a href="https://github.com/feiyoung/PRECAST_Analysis/issues/1" class="external-link">issue</a> reported by Boyi Guo. Specifically, the assay name “RNA” used in functions <code><a href="reference/CreateiDRSCObject.html">CreatePRECASTObject()</a></code> and <code><a href="reference/PRECAST.html">PRECAST()</a></code> is replaced by the default assay using <code>DefaultAssay</code> function in Seurat.</p></li>
<li><p>Provide more detailed help file for <code><a href="reference/CreateiDRSCObject.html">CreatePRECASTObject()</a></code> function. Users can use <code><a href="reference/CreateiDRSCObject.html">?CreatePRECASTObject</a></code> in Rstudio to access the help file. In detail, seuList is a list with Seurat object as component, and each Seurat object at least includes the raw expression count matrix, and spatial coordinates in metadata for each data batch, where the spatial coordinates information must be saved in the metadata of Seurat, named “row” and “col” for each data batch. See the help file for more details.</p></li>
<li><p>Add the <a href="https://github.com/feiyoung/PRECAST/tree/main/vignettes_data" class="external-link">data</a> used in <a href="https://feiyoung.github.io/PRECAST/index.html" class="external-link">Package Website</a>.</p></li>
<li><p>Add the wrapper functions for different common-used R objects in the spatial transcriptomics, such as spatialExperiment, etc; see the functions <code>spe2seurat</code>, <code>spe2seuratList</code>,<code>seu2seuList</code> and their help file in our developed <a href="https://github.com/feiyoung/SRTtools" class="external-link">SRTtools R package</a>. This R package is designed to provide auxiliary functions for the serirs of our developed methods, such as <a href="https://github.com/Shufeyangyi2015310117/SC.MEB" class="external-link">SC-MEB</a>, <a href="https://github.com/feiyoung//DR.SC" class="external-link">DR-SC</a> and <a href="https://github.com/feiyoung/PRECAST" class="external-link">PRECAST</a>.</p></li>
</ul>
</div>

  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://cloud.r-project.org/package=PRECAST" class="external-link">View on CRAN</a></li>
<li><a href="https://github.com/feiyoung/PRECAST/" class="external-link">Browse source code</a></li>
<li><a href="https://github.com/feiyoung/PRECAST/issues" class="external-link">Report a bug</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="https://www.r-project.org/Licenses/GPL-3" class="external-link">GPL-3</a></li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing PRECAST</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Wei Liu <br><small class="roles"> Author, maintainer </small>  </li>
<li>Yi Yang <br><small class="roles"> Author </small>  </li>
<li>Jin Liu <br><small class="roles"> Author </small>  </li>
</ul>
</div>



  </div>
</div>


      <footer><div class="copyright">
  <p></p>
<p>Developed by Wei Liu, Yi Yang, Jin Liu.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
