<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>PRECAST: Human Breast Cancer Data Analysis • PRECAST</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="PRECAST: Human Breast Cancer Data Analysis">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">


    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">PRECAST</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.7</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/PRECAST.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/PRECAST.BreastCancer.html">PRECAST: Human Breast Cancer Data Analysis</a>
    </li>
    <li>
      <a href="../articles/PRECAST.DLPFC.html">PRECAST: DLPFC Single Sample Analysis</a>
    </li>
    <li>
      <a href="../articles/PRECAST.DLPFC4.html">PRECAST: Four DLPFC Sample Analysis</a>
    </li>
    <li>
      <a href="../articles/PRECAST.Simu.html">PRECAST: simulation</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/feiyoung/PRECAST/" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>PRECAST: Human Breast Cancer Data Analysis</h1>
                        <h4 data-toc-skip class="author">Wei Liu</h4>
            
            <h4 data-toc-skip class="date">2025-09-29</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/feiyoung/PRECAST/blob/HEAD/vignettes/PRECAST.BreastCancer.Rmd" class="external-link"><code>vignettes/PRECAST.BreastCancer.Rmd</code></a></small>
      <div class="hidden name"><code>PRECAST.BreastCancer.Rmd</code></div>

    </div>

    
    
<p>This vignette introduces the PRECAST workflow for the analysis of
integrating multiple spatial transcriptomics datasets. The workflow
consists of three steps</p>
<ul>
<li>Independent preprocessing and model setting</li>
<li>Probabilistic embedding, clustering and alignment using PRECAST
model</li>
<li>Downstream analysis (i.e. visualization of clusters and embeddings,
combined differential expression analysis)</li>
</ul>
<p>We demonstrate the use of PRECAST to two sliced human breast cancer
Visium data that are <a href="https://github.com/feiyoung/PRECAST/tree/main/vignettes_data" class="external-link">here</a>,
which can be downloaded to the current working path by the following
command:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">githubURL</span> <span class="op">&lt;-</span> <span class="st">"https://github.com/feiyoung/PRECAST/blob/main/vignettes_data/bc2.rda?raw=true"</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/download.file.html" class="external-link">download.file</a></span><span class="op">(</span><span class="va">githubURL</span>, <span class="st">"bc2.rda"</span>, mode <span class="op">=</span> <span class="st">"wb"</span><span class="op">)</span></span></code></pre></div>
<p>Then load to R</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/load.html" class="external-link">load</a></span><span class="op">(</span><span class="st">"bc2.rda"</span><span class="op">)</span></span></code></pre></div>
<details><summary><strong>Download data from 10X: another method to access data</strong>
</summary><p>This data is also available at 10X genomics data website:</p>
<ul>
<li>Section1: <a href="https://support.10xgenomics.com/spatial-gene-expression/datasets/1.1.0/V1_Breast_Cancer_Block_A_Section_1" class="external-link uri">https://support.10xgenomics.com/spatial-gene-expression/datasets/1.1.0/V1_Breast_Cancer_Block_A_Section_1</a>
</li>
<li>Section2: <a href="https://support.10xgenomics.com/spatial-gene-expression/datasets/1.1.0/V1_Breast_Cancer_Block_A_Section_2" class="external-link uri">https://support.10xgenomics.com/spatial-gene-expression/datasets/1.1.0/V1_Breast_Cancer_Block_A_Section_2</a>
</li>
</ul>
<p>Users require the two folders for each dataset: spatial and
filtered_feature_bc_matrix. Then the data can be read by the following
commond.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dir.file</span> <span class="op">&lt;-</span> <span class="st">"Section"</span>  <span class="co">## the folders Section1 and Section2, and each includes two folders spatial and filtered_feature_bc_matrix</span></span>
<span><span class="va">seuList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">r</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">message</a></span><span class="op">(</span><span class="st">"r = "</span>, <span class="va">r</span><span class="op">)</span></span>
<span>    <span class="va">seuList</span><span class="op">[[</span><span class="va">r</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu">DR.SC</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/DR.SC/man/read10XVisium.html" class="external-link">read10XVisium</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">dir.file</span>, <span class="va">r</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="va">bc2</span> <span class="op">&lt;-</span> <span class="va">seuList</span></span></code></pre></div>
</details><p>The package can be loaded with the command:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/feiyoung/PRECAST" class="external-link">PRECAST</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span></span></code></pre></div>
<p>View human breast cancer Visium data from
<code>DataPRECAST</code></p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">bc2</span>  <span class="co">## a list including two Seurat object</span></span>
<span><span class="co">#&gt; [[1]]</span></span>
<span><span class="co">#&gt; An object of class Seurat </span></span>
<span><span class="co">#&gt; 36601 features across 3798 samples within 1 assay </span></span>
<span><span class="co">#&gt; Active assay: RNA (36601 features, 0 variable features)</span></span>
<span><span class="co">#&gt;  2 layers present: counts, data</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; [[2]]</span></span>
<span><span class="co">#&gt; An object of class Seurat </span></span>
<span><span class="co">#&gt; 36601 features across 3987 samples within 1 assay </span></span>
<span><span class="co">#&gt; Active assay: RNA (36601 features, 0 variable features)</span></span>
<span><span class="co">#&gt;  2 layers present: counts, data</span></span></code></pre></div>
<p>Check the content in <code>bc2</code></p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">bc2</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="co">#&gt;                       orig.ident nCount_RNA nFeature_RNA               spot</span></span>
<span><span class="co">#&gt; CACGATTGGTCGTTAA-1 SeuratProject       6853         3026 CACGATTGGTCGTTAA-1</span></span>
<span><span class="co">#&gt; GGTTGTATCGTGAAAT-1 SeuratProject       8679         3646 GGTTGTATCGTGAAAT-1</span></span>
<span><span class="co">#&gt; TCTTATGGGTAGTACC-1 SeuratProject       5093         2449 TCTTATGGGTAGTACC-1</span></span>
<span><span class="co">#&gt; TACAAGCTGTTCACTG-1 SeuratProject       9432         3708 TACAAGCTGTTCACTG-1</span></span>
<span><span class="co">#&gt; GTATCTTGTTGCTCAC-1 SeuratProject       8610         3512 GTATCTTGTTGCTCAC-1</span></span>
<span><span class="co">#&gt; ATACCAGGTGAGCGAT-1 SeuratProject       9974         3901 ATACCAGGTGAGCGAT-1</span></span>
<span><span class="co">#&gt; CCTAAACAGGGTCCGT-1 SeuratProject       7232         3173 CCTAAACAGGGTCCGT-1</span></span>
<span><span class="co">#&gt; ATGGTGCTCAAAGCCA-1 SeuratProject       6025         2867 ATGGTGCTCAAAGCCA-1</span></span>
<span><span class="co">#&gt; CAAATGCGGAGTGTTC-1 SeuratProject       8822         3678 CAAATGCGGAGTGTTC-1</span></span>
<span><span class="co">#&gt; CGTGCCCGACATTTGT-1 SeuratProject       6364         2932 CGTGCCCGACATTTGT-1</span></span>
<span><span class="co">#&gt;                    in_tissue row col imagerow imagecol</span></span>
<span><span class="co">#&gt; CACGATTGGTCGTTAA-1         1   0  50     4046    10350</span></span>
<span><span class="co">#&gt; GGTTGTATCGTGAAAT-1         1   1  51     4284    10486</span></span>
<span><span class="co">#&gt; TCTTATGGGTAGTACC-1         1   0  52     4047    10623</span></span>
<span><span class="co">#&gt; TACAAGCTGTTCACTG-1         1   1  53     4285    10759</span></span>
<span><span class="co">#&gt; GTATCTTGTTGCTCAC-1         1   0  54     4047    10896</span></span>
<span><span class="co">#&gt; ATACCAGGTGAGCGAT-1         1   1  55     4285    11032</span></span>
<span><span class="co">#&gt; CCTAAACAGGGTCCGT-1         1   0  56     4048    11169</span></span>
<span><span class="co">#&gt; ATGGTGCTCAAAGCCA-1         1   1  57     4286    11305</span></span>
<span><span class="co">#&gt; CAAATGCGGAGTGTTC-1         1   0  58     4048    11442</span></span>
<span><span class="co">#&gt; CGTGCCCGACATTTGT-1         1   1  59     4286    11578</span></span></code></pre></div>
<div class="section level2">
<h2 id="create-a-precastobject-object">Create a PRECASTObject object<a class="anchor" aria-label="anchor" href="#create-a-precastobject-object"></a>
</h2>
<p>We show how to create a PRECASTObject object step by step. First, we
create a Seurat list object using the count matrix and meta data of each
data batch. Although <code>bc2</code> is a prepared Seurat list object,
we re-create it to show the details of the Seurat list object. At the
same time, check the meta data that must include the spatial coordinates
named “row” and “col”, respectively. If the names are not, they are
required to rename them.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Get the gene-by-spot read count matrices countList &lt;- lapply(bc2, function(x)</span></span>
<span><span class="co">## x[['RNA']]@counts)</span></span>
<span><span class="va">countList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">bc2</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">assay</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.github.io/seurat-object/reference/DefaultAssay.html" class="external-link">DefaultAssay</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://satijalab.github.io/seurat-object/reference/AssayData.html" class="external-link">GetAssayData</a></span><span class="op">(</span><span class="va">x</span>, assay <span class="op">=</span> <span class="va">assay</span>, slot <span class="op">=</span> <span class="st">"counts"</span><span class="op">)</span></span>
<span></span>
<span><span class="op">}</span><span class="op">)</span></span>
<span></span>
<span><span class="va">M</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">countList</span><span class="op">)</span></span>
<span><span class="co">## Get the meta data of each spot for each data batch</span></span>
<span><span class="va">metadataList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">bc2</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="va">x</span><span class="op">@</span><span class="va">meta.data</span><span class="op">)</span></span>
<span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">r</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="va">M</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">meta_data</span> <span class="op">&lt;-</span> <span class="va">metadataList</span><span class="op">[[</span><span class="va">r</span><span class="op">]</span><span class="op">]</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/all.html" class="external-link">all</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"row"</span>, <span class="st">"col"</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">meta_data</span><span class="op">)</span><span class="op">)</span>  <span class="co">## the names are correct!</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">meta_data</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"row"</span>, <span class="st">"col"</span><span class="op">)</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span></span>
<span><span class="co">## ensure the row.names of metadata in metaList are the same as that of colnames count matrix</span></span>
<span><span class="co">## in countList</span></span>
<span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">r</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="va">M</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/row.names.html" class="external-link">row.names</a></span><span class="op">(</span><span class="va">metadataList</span><span class="op">[[</span><span class="va">r</span><span class="op">]</span><span class="op">]</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">countList</span><span class="op">[[</span><span class="va">r</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span></span>
<span><span class="co">## Create the Seurat list object</span></span>
<span></span>
<span><span class="va">seuList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">r</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="va">M</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">seuList</span><span class="op">[[</span><span class="va">r</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.github.io/seurat-object/reference/CreateSeuratObject.html" class="external-link">CreateSeuratObject</a></span><span class="op">(</span>counts <span class="op">=</span> <span class="va">countList</span><span class="op">[[</span><span class="va">r</span><span class="op">]</span><span class="op">]</span>, meta.data <span class="op">=</span> <span class="va">metadataList</span><span class="op">[[</span><span class="va">r</span><span class="op">]</span><span class="op">]</span>, project <span class="op">=</span> <span class="st">"BreastCancerPRECAST"</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="va">bc2</span> <span class="op">&lt;-</span> <span class="va">seuList</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/rm.html" class="external-link">rm</a></span><span class="op">(</span><span class="va">seuList</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">meta_data</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"row"</span>, <span class="st">"col"</span><span class="op">)</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<div class="section level3">
<h3 id="prepare-the-precastobject-with-preprocessing-step-">Prepare the PRECASTObject with preprocessing step.<a class="anchor" aria-label="anchor" href="#prepare-the-precastobject-with-preprocessing-step-"></a>
</h3>
<p>Next, we use <code><a href="../reference/CreateiDRSCObject.html">CreatePRECASTObject()</a></code> to create a
PRECASTObject based on the Seurat list object <code>bc2</code>. This
function will do three things:</p>
<ul>
<li><ol style="list-style-type: decimal">
<li>Filter low-quality spots and genes, controlled by the arguments
<code>premin.features</code> and <code>premin.spots</code>,
respectively; the spots are retained in raw data (bc2) with at least
premin.features number of nonzero-count features (genes), and the genes
are retained in raw data (bc2) with at least <code>premin.spots</code>
number of spots. To ease presentation, we denote the filtered Seurat
list object as bc2_filter1.</li>
</ol></li>
<li><ol start="2" style="list-style-type: decimal">
<li>Select the top 2,000 variable genes (by setting
<code>gene.number=2000</code>) for each data batch using
<code>FindSVGs()</code> function in <code>DR.SC</code> package for
spatially variable genes or <code><a href="https://satijalab.org/seurat/reference/FindVariableFeatures.html" class="external-link">FindVariableFeatures()</a></code> function
in <code>Seurat</code> package for highly variable genes. Next, we
prioritized genes based on the number of times they were selected as
variable genes in all samples and chose the top 2,000 genes. Then denote
the Seurat list object as bc2_filter2, where only 2,000 genes are
retained.</li>
</ol></li>
<li><ol start="3" style="list-style-type: decimal">
<li>Conduct strict quality control for bc2_filter2 by filtering spots
and genes, controlled by the arguments <code>postmin.features</code> and
<code>postmin.spots</code>, respectively; the spots are retained with at
least <code>post.features</code> nonzero counts across genes; the
features (genes) are retained with at least <code>postmin.spots</code>
number of nonzero-count spots. Usually, no genes are filltered because
these genes are variable genes.</li>
</ol></li>
</ul>
<p>If the argument <code>customGenelist</code> is not <code>NULL</code>,
then this function only does (3) not (1) and (2). User can retain the
raw seurat list object by setting
<code>rawData.preserve = TRUE</code>.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Create PRECASTObject.</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">2022</span><span class="op">)</span></span>
<span><span class="va">PRECASTObj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/CreateiDRSCObject.html">CreatePRECASTObject</a></span><span class="op">(</span><span class="va">bc2</span>, project <span class="op">=</span> <span class="st">"BC2"</span>, gene.number <span class="op">=</span> <span class="fl">2000</span>, selectGenesMethod <span class="op">=</span> <span class="st">"SPARK-X"</span>,</span>
<span>    premin.spots <span class="op">=</span> <span class="fl">20</span>, premin.features <span class="op">=</span> <span class="fl">20</span>, postmin.spots <span class="op">=</span> <span class="fl">1</span>, postmin.features <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## User can retain the raw seuList by the following commond.  PRECASTObj &lt;-</span></span>
<span><span class="co">## CreatePRECASTObject(seuList, customGenelist=row.names(seuList[[1]]), rawData.preserve =</span></span>
<span><span class="co">## TRUE)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="add-the-model-setting">Add the model setting<a class="anchor" aria-label="anchor" href="#add-the-model-setting"></a>
</h3>
<p>Add adjacency matrix list and parameter setting of PRECAST. More
model setting parameters can be found in .</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## check the number of genes/features after filtering step</span></span>
<span><span class="va">PRECASTObj</span><span class="op">@</span><span class="va">seulist</span></span>
<span><span class="co">#&gt; [[1]]</span></span>
<span><span class="co">#&gt; An object of class Seurat </span></span>
<span><span class="co">#&gt; 2000 features across 3798 samples within 1 assay </span></span>
<span><span class="co">#&gt; Active assay: RNA (2000 features, 1830 variable features)</span></span>
<span><span class="co">#&gt;  2 layers present: counts, data</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; [[2]]</span></span>
<span><span class="co">#&gt; An object of class Seurat </span></span>
<span><span class="co">#&gt; 2000 features across 3987 samples within 1 assay </span></span>
<span><span class="co">#&gt; Active assay: RNA (2000 features, 1850 variable features)</span></span>
<span><span class="co">#&gt;  2 layers present: counts, data</span></span>
<span></span>
<span><span class="co">## seuList is null since the default value `rawData.preserve` is FALSE.</span></span>
<span><span class="va">PRECASTObj</span><span class="op">@</span><span class="va">seuList</span></span>
<span><span class="co">#&gt; NULL</span></span>
<span></span>
<span><span class="co">## Add adjacency matrix list for a PRECASTObj object to prepare for PRECAST model fitting.</span></span>
<span><span class="va">PRECASTObj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/AddAdjList.html">AddAdjList</a></span><span class="op">(</span><span class="va">PRECASTObj</span>, platform <span class="op">=</span> <span class="st">"Visium"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Add a model setting in advance for a PRECASTObj object. verbose =TRUE helps outputing the</span></span>
<span><span class="co">## information in the algorithm.</span></span>
<span><span class="va">PRECASTObj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/AddParSetting.html">AddParSetting</a></span><span class="op">(</span><span class="va">PRECASTObj</span>, Sigma_equal <span class="op">=</span> <span class="cn">FALSE</span>, verbose <span class="op">=</span> <span class="cn">TRUE</span>, maxIter <span class="op">=</span> <span class="fl">30</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="fit-precast-using-this-data">Fit PRECAST using this data<a class="anchor" aria-label="anchor" href="#fit-precast-using-this-data"></a>
</h2>
<div class="section level3">
<h3 id="fit-precast">Fit PRECAST<a class="anchor" aria-label="anchor" href="#fit-precast"></a>
</h3>
<p>For function <code>PRECAST</code>, users can specify the number of
clusters
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>K</mi><annotation encoding="application/x-tex">K</annotation></semantics></math>
or set <code>K</code> to be an integer vector by using modified
BIC(MBIC) to determine
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>K</mi><annotation encoding="application/x-tex">K</annotation></semantics></math>.
First, we try using user-specified number of clusters. For convenience,
we give the selected number of clusters by MBIC (K=14).</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">### Given K</span></span>
<span><span class="va">PRECASTObj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/PRECAST.html">PRECAST</a></span><span class="op">(</span><span class="va">PRECASTObj</span>, K <span class="op">=</span> <span class="fl">14</span><span class="op">)</span></span>
<span><span class="co">#&gt; iter = 2, loglik= 2975971.500000, dloglik=1.001386 </span></span>
<span><span class="co">#&gt; iter = 3, loglik= 3027712.750000, dloglik=0.017386 </span></span>
<span><span class="co">#&gt; iter = 4, loglik= 3047752.250000, dloglik=0.006619 </span></span>
<span><span class="co">#&gt; iter = 5, loglik= 3058068.750000, dloglik=0.003385 </span></span>
<span><span class="co">#&gt; iter = 6, loglik= 3064000.000000, dloglik=0.001940 </span></span>
<span><span class="co">#&gt; iter = 7, loglik= 3067513.000000, dloglik=0.001147 </span></span>
<span><span class="co">#&gt; iter = 8, loglik= 3069745.000000, dloglik=0.000728 </span></span>
<span><span class="co">#&gt; iter = 9, loglik= 3071316.250000, dloglik=0.000512 </span></span>
<span><span class="co">#&gt; iter = 10, loglik= 3072378.500000, dloglik=0.000346 </span></span>
<span><span class="co">#&gt; iter = 11, loglik= 3073222.750000, dloglik=0.000275 </span></span>
<span><span class="co">#&gt; iter = 12, loglik= 3073758.500000, dloglik=0.000174 </span></span>
<span><span class="co">#&gt; iter = 13, loglik= 3074317.500000, dloglik=0.000182 </span></span>
<span><span class="co">#&gt; iter = 14, loglik= 3074639.000000, dloglik=0.000105 </span></span>
<span><span class="co">#&gt; iter = 15, loglik= 3075017.750000, dloglik=0.000123 </span></span>
<span><span class="co">#&gt; iter = 16, loglik= 3075328.250000, dloglik=0.000101 </span></span>
<span><span class="co">#&gt; iter = 17, loglik= 3075638.500000, dloglik=0.000101 </span></span>
<span><span class="co">#&gt; iter = 18, loglik= 3075812.750000, dloglik=0.000057 </span></span>
<span><span class="co">#&gt; iter = 19, loglik= 3076067.750000, dloglik=0.000083 </span></span>
<span><span class="co">#&gt; iter = 20, loglik= 3076207.750000, dloglik=0.000046 </span></span>
<span><span class="co">#&gt; iter = 21, loglik= 3076425.750000, dloglik=0.000071 </span></span>
<span><span class="co">#&gt; iter = 22, loglik= 3076523.250000, dloglik=0.000032 </span></span>
<span><span class="co">#&gt; iter = 23, loglik= 3076621.250000, dloglik=0.000032 </span></span>
<span><span class="co">#&gt; iter = 24, loglik= 3076780.500000, dloglik=0.000052 </span></span>
<span><span class="co">#&gt; iter = 25, loglik= 3076890.750000, dloglik=0.000036 </span></span>
<span><span class="co">#&gt; iter = 26, loglik= 3076923.750000, dloglik=0.000011 </span></span>
<span><span class="co">#&gt; iter = 27, loglik= 3077042.000000, dloglik=0.000038 </span></span>
<span><span class="co">#&gt; iter = 28, loglik= 3077170.000000, dloglik=0.000042 </span></span>
<span><span class="co">#&gt; iter = 29, loglik= 3077211.500000, dloglik=0.000013 </span></span>
<span><span class="co">#&gt; iter = 30, loglik= 3077308.250000, dloglik=0.000031</span></span></code></pre></div>
<p>Select a best model if
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>K</mi><annotation encoding="application/x-tex">K</annotation></semantics></math>
is an integer vector. Even if
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>K</mi><annotation encoding="application/x-tex">K</annotation></semantics></math>
is a scalar, this step is also neccessary to re-organize the results in
<code>PRECASTObj</code> object.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## backup the fitting results in resList</span></span>
<span><span class="va">resList</span> <span class="op">&lt;-</span> <span class="va">PRECASTObj</span><span class="op">@</span><span class="va">resList</span></span>
<span><span class="va">PRECASTObj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/selectModel.html">SelectModel</a></span><span class="op">(</span><span class="va">PRECASTObj</span><span class="op">)</span></span></code></pre></div>
<p>Integrate the two samples using the <code>IntegrateSpaData</code>
function. For computational efficiency, this function exclusively
integrates the variable genes. Specifically, in cases where users do not
specify the <code>PRECASTObj@seuList</code> or <code>seuList</code>
argument within the <code>IntegrateSpaData</code> function, it
automatically focuses on integrating only the variable genes. The
default setting for <code>PRECASTObj@seuList</code> is <code>NULL</code>
when <code>rawData.preserve</code> in <code>CreatePRECASTObject</code>
is set to <code>FALSE</code>. For instance:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">PRECASTObj</span><span class="op">@</span><span class="va">seuList</span><span class="op">)</span></span>
<span><span class="co">#&gt; NULL</span></span>
<span><span class="va">seuInt</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/IntegrateSpaData.html">IntegrateSpaData</a></span><span class="op">(</span><span class="va">PRECASTObj</span>, species <span class="op">=</span> <span class="st">"Human"</span><span class="op">)</span></span>
<span><span class="va">seuInt</span></span>
<span><span class="co">#&gt; An object of class Seurat </span></span>
<span><span class="co">#&gt; 2000 features across 7785 samples within 1 assay </span></span>
<span><span class="co">#&gt; Active assay: PRE_CAST (2000 features, 0 variable features)</span></span>
<span><span class="co">#&gt;  2 layers present: counts, data</span></span>
<span><span class="co">#&gt;  2 dimensional reductions calculated: PRECAST, position</span></span>
<span><span class="co">## The low-dimensional embeddings obtained by PRECAST are saved in PRECAST reduction slot.</span></span></code></pre></div>
<details><summary><strong>Integrating all genes</strong>
</summary><p>There are two ways to use <code>IntegrateSpaData</code> integrating
all genes, which will require more memory. We recommand running for all
genes on server. The first one is to set value for
<code>PRECASTObj@seuList</code>.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## assign the raw Seurat list object to it For illustration, we generate a new seuList with</span></span>
<span><span class="co">## more genes; For integrating all genes, users can set `seuList &lt;- bc2`.</span></span>
<span><span class="va">genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/row.names.html" class="external-link">row.names</a></span><span class="op">(</span><span class="va">PRECASTObj</span><span class="op">@</span><span class="va">seulist</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/row.names.html" class="external-link">row.names</a></span><span class="op">(</span><span class="va">bc2</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="va">seuList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">bc2</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="va">x</span><span class="op">[</span><span class="va">genes</span>, <span class="op">]</span><span class="op">)</span></span>
<span><span class="va">PRECASTObj</span><span class="op">@</span><span class="va">seuList</span> <span class="op">&lt;-</span> <span class="va">seuList</span>  <span class="co"># </span></span>
<span><span class="va">seuInt</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/IntegrateSpaData.html">IntegrateSpaData</a></span><span class="op">(</span><span class="va">PRECASTObj</span>, species <span class="op">=</span> <span class="st">"Human"</span><span class="op">)</span></span>
<span><span class="va">seuInt</span></span></code></pre></div>
<p>The second method is to set a value for the argument
<code>seuList</code>:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">PRECASTObj</span><span class="op">@</span><span class="va">seuList</span> <span class="op">&lt;-</span> <span class="cn">NULL</span></span>
<span><span class="co">## At the same time, we can set subsampling to speed up the computation.</span></span>
<span><span class="va">seuInt</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/IntegrateSpaData.html">IntegrateSpaData</a></span><span class="op">(</span><span class="va">PRECASTObj</span>, species <span class="op">=</span> <span class="st">"Human"</span>, seuList <span class="op">=</span> <span class="va">seuList</span>, subsample_rate <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span></span>
<span><span class="va">seuInt</span></span></code></pre></div>
</details><p>First, user can choose a beautiful color schema using
<code><a href="../reference/chooseColors.html">chooseColors()</a></code>.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cols_cluster</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/chooseColors.html">chooseColors</a></span><span class="op">(</span>palettes_name <span class="op">=</span> <span class="st">"Classic 20"</span>, n_colors <span class="op">=</span> <span class="fl">14</span>, plot_colors <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p>Show the spatial scatter plot for clusters</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">p12</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/SpaPlot.html">SpaPlot</a></span><span class="op">(</span><span class="va">seuInt</span>, item <span class="op">=</span> <span class="st">"cluster"</span>, batch <span class="op">=</span> <span class="cn">NULL</span>, point_size <span class="op">=</span> <span class="fl">1</span>, cols <span class="op">=</span> <span class="va">cols_cluster</span>, combine <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    nrow.legend <span class="op">=</span> <span class="fl">7</span><span class="op">)</span></span>
<span><span class="va">p12</span></span></code></pre></div>
<p><img src="PRECAST.BreastCancer_files/figure-html/unnamed-chunk-17-1.png" width="864"></p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># users can plot each sample by setting combine=FALSE</span></span></code></pre></div>
<p>Users can re-plot the above figures for specific need by returning a
ggplot list object. For example, we plot the spatial heatmap using a
common legend.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pList</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/SpaPlot.html">SpaPlot</a></span><span class="op">(</span><span class="va">seuInt</span>, item <span class="op">=</span> <span class="st">"cluster"</span>, batch <span class="op">=</span> <span class="cn">NULL</span>, point_size <span class="op">=</span> <span class="fl">1</span>, cols <span class="op">=</span> <span class="va">cols_cluster</span>, combine <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>    nrow.legend <span class="op">=</span> <span class="fl">7</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/drawFigs.html">drawFigs</a></span><span class="op">(</span><span class="va">pList</span>, layout.dim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span><span class="op">)</span>, common.legend <span class="op">=</span> <span class="cn">TRUE</span>, legend.position <span class="op">=</span> <span class="st">"right"</span>, align <span class="op">=</span> <span class="st">"hv"</span><span class="op">)</span></span></code></pre></div>
<p><img src="PRECAST.BreastCancer_files/figure-html/unnamed-chunk-18-1.png" width="720"></p>
<p>Show the spatial UMAP/tNSE RGB plot to illustrate the performance in
extracting features.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seuInt</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/AddUMAP.html">AddUMAP</a></span><span class="op">(</span><span class="va">seuInt</span><span class="op">)</span></span>
<span><span class="va">p13</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/SpaPlot.html">SpaPlot</a></span><span class="op">(</span><span class="va">seuInt</span>, batch <span class="op">=</span> <span class="cn">NULL</span>, item <span class="op">=</span> <span class="st">"RGB_UMAP"</span>, point_size <span class="op">=</span> <span class="fl">2</span>, combine <span class="op">=</span> <span class="cn">TRUE</span>, text_size <span class="op">=</span> <span class="fl">15</span><span class="op">)</span></span>
<span><span class="va">p13</span></span></code></pre></div>
<p><img src="PRECAST.BreastCancer_files/figure-html/unnamed-chunk-19-1.png" width="547.2"></p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># seuInt &lt;- AddTSNE(seuInt) SpaPlot(seuInt, batch=NULL,item='RGB_TSNE',point_size=2,</span></span>
<span><span class="co"># combine=T, text_size=15)</span></span></code></pre></div>
<p>Show the tSNE plot based on the extracted features from PRECAST to
check the performance of integration.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seuInt</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/AddTSNE.html">AddTSNE</a></span><span class="op">(</span><span class="va">seuInt</span>, n_comp <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span>
<span><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/dimPlot.html">dimPlot</a></span><span class="op">(</span><span class="va">seuInt</span>, item <span class="op">=</span> <span class="st">"cluster"</span>, point_size <span class="op">=</span> <span class="fl">0.5</span>, font_family <span class="op">=</span> <span class="st">"serif"</span>, cols <span class="op">=</span> <span class="va">cols_cluster</span>,</span>
<span>    border_col <span class="op">=</span> <span class="st">"gray10"</span>, nrow.legend <span class="op">=</span> <span class="fl">14</span>, legend_pos <span class="op">=</span> <span class="st">"right"</span><span class="op">)</span>  <span class="co"># Times New Roman</span></span>
<span><span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/dimPlot.html">dimPlot</a></span><span class="op">(</span><span class="va">seuInt</span>, item <span class="op">=</span> <span class="st">"batch"</span>, point_size <span class="op">=</span> <span class="fl">0.5</span>, font_family <span class="op">=</span> <span class="st">"serif"</span>, legend_pos <span class="op">=</span> <span class="st">"right"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/drawFigs.html">drawFigs</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">p1</span>, <span class="va">p2</span><span class="op">)</span>, layout.dim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span><span class="op">)</span>, legend.position <span class="op">=</span> <span class="st">"right"</span>, align <span class="op">=</span> <span class="st">"hv"</span><span class="op">)</span></span></code></pre></div>
<p><img src="PRECAST.BreastCancer_files/figure-html/unnamed-chunk-20-1.png" width="1152"></p>
<p>Combined differential expression analysis</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span></span>
<span><span class="va">dat_deg</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindAllMarkers.html" class="external-link">FindAllMarkers</a></span><span class="op">(</span><span class="va">seuInt</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span><span class="va">n</span> <span class="op">&lt;-</span> <span class="fl">10</span></span>
<span><span class="va">dat_deg</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/dplyr/man/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">cluster</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/dplyr/man/top_n.html" class="external-link">top_n</a></span><span class="op">(</span>n <span class="op">=</span> <span class="va">n</span>, wt <span class="op">=</span> <span class="va">avg_log2FC</span><span class="op">)</span> <span class="op">-&gt;</span> <span class="va">top10</span></span>
<span></span>
<span><span class="va">seuInt</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/ScaleData.html" class="external-link">ScaleData</a></span><span class="op">(</span><span class="va">seuInt</span><span class="op">)</span></span>
<span><span class="va">seus</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">seuInt</span>, downsample <span class="op">=</span> <span class="fl">400</span><span class="op">)</span></span></code></pre></div>
<p>Plot DE genes’ heatmap for each spatial domain identified by
PRECAST.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">color_id</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/levels.html" class="external-link">levels</a></span><span class="op">(</span><span class="fu"><a href="https://satijalab.github.io/seurat-object/reference/Idents.html" class="external-link">Idents</a></span><span class="op">(</span><span class="va">seus</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="co">## HeatMap</span></span>
<span><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/doHeatmap.html">doHeatmap</a></span><span class="op">(</span><span class="va">seus</span>, features <span class="op">=</span> <span class="va">top10</span><span class="op">$</span><span class="va">gene</span>, cell_label <span class="op">=</span> <span class="st">"Domain"</span>, grp_label <span class="op">=</span> <span class="cn">F</span>, grp_color <span class="op">=</span> <span class="va">cols_cluster</span><span class="op">[</span><span class="va">color_id</span><span class="op">]</span>,</span>
<span>    pt_size <span class="op">=</span> <span class="fl">6</span>, slot <span class="op">=</span> <span class="st">"scale.data"</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://rdrr.io/pkg/ggplot2/man/theme.html" class="external-link">theme</a></span><span class="op">(</span>legend.text <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/ggplot2/man/element.html" class="external-link">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">10</span><span class="op">)</span>, legend.title <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/ggplot2/man/element.html" class="external-link">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">13</span>,</span>
<span>    face <span class="op">=</span> <span class="st">"bold"</span><span class="op">)</span>, axis.text.y <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/ggplot2/man/element.html" class="external-link">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">5</span>, face <span class="op">=</span> <span class="st">"italic"</span>, family <span class="op">=</span> <span class="st">"serif"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">p1</span></span></code></pre></div>
<p><img src="PRECAST.BreastCancer_files/figure-html/unnamed-chunk-22-1.png" width="864"></p>
<details><summary><strong>Session Info</strong>
</summary><div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.4.1 (2024-06-14 ucrt)</span></span>
<span><span class="co">#&gt; Platform: x86_64-w64-mingw32/x64</span></span>
<span><span class="co">#&gt; Running under: Windows 11 x64 (build 26100)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt; [1] LC_COLLATE=Chinese (Simplified)_China.utf8 </span></span>
<span><span class="co">#&gt; [2] LC_CTYPE=Chinese (Simplified)_China.utf8   </span></span>
<span><span class="co">#&gt; [3] LC_MONETARY=Chinese (Simplified)_China.utf8</span></span>
<span><span class="co">#&gt; [4] LC_NUMERIC=C                               </span></span>
<span><span class="co">#&gt; [5] LC_TIME=Chinese (Simplified)_China.utf8    </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: Asia/Shanghai</span></span>
<span><span class="co">#&gt; tzcode source: internal</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] parallel  stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">#&gt; [8] base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt; [1] ggplot2_3.5.2      dplyr_1.1.4        irlba_2.3.5.1      Matrix_1.7-0      </span></span>
<span><span class="co">#&gt; [5] Seurat_5.1.0       SeuratObject_5.0.2 sp_2.1-4           PRECAST_1.7       </span></span>
<span><span class="co">#&gt; [9] gtools_3.9.5      </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;   [1] RcppAnnoy_0.0.22            splines_4.4.1              </span></span>
<span><span class="co">#&gt;   [3] later_1.3.2                 tibble_3.2.1               </span></span>
<span><span class="co">#&gt;   [5] polyclip_1.10-7             fastDummies_1.7.4          </span></span>
<span><span class="co">#&gt;   [7] lifecycle_1.0.4             rstatix_0.7.2              </span></span>
<span><span class="co">#&gt;   [9] globals_0.16.3              lattice_0.22-6             </span></span>
<span><span class="co">#&gt;  [11] MASS_7.3-60.2               backports_1.5.0            </span></span>
<span><span class="co">#&gt;  [13] magrittr_2.0.3              limma_3.58.1               </span></span>
<span><span class="co">#&gt;  [15] plotly_4.10.4               sass_0.4.9                 </span></span>
<span><span class="co">#&gt;  [17] rmarkdown_2.28              jquerylib_0.1.4            </span></span>
<span><span class="co">#&gt;  [19] yaml_2.3.10                 httpuv_1.6.15              </span></span>
<span><span class="co">#&gt;  [21] sctransform_0.4.1           spam_2.10-0                </span></span>
<span><span class="co">#&gt;  [23] spatstat.sparse_3.1-0       reticulate_1.39.0          </span></span>
<span><span class="co">#&gt;  [25] cowplot_1.1.3               pbapply_1.7-2              </span></span>
<span><span class="co">#&gt;  [27] RColorBrewer_1.1-3          abind_1.4-8                </span></span>
<span><span class="co">#&gt;  [29] zlibbioc_1.50.0             Rtsne_0.17                 </span></span>
<span><span class="co">#&gt;  [31] GenomicRanges_1.56.2        presto_1.0.0               </span></span>
<span><span class="co">#&gt;  [33] purrr_1.0.2                 BiocGenerics_0.50.0        </span></span>
<span><span class="co">#&gt;  [35] GenomeInfoDbData_1.2.12     IRanges_2.38.1             </span></span>
<span><span class="co">#&gt;  [37] S4Vectors_0.42.1            ggrepel_0.9.6              </span></span>
<span><span class="co">#&gt;  [39] listenv_0.9.1               spatstat.utils_3.1-0       </span></span>
<span><span class="co">#&gt;  [41] goftest_1.2-3               RSpectra_0.16-2            </span></span>
<span><span class="co">#&gt;  [43] spatstat.random_3.3-2       fitdistrplus_1.2-1         </span></span>
<span><span class="co">#&gt;  [45] parallelly_1.38.0           pkgdown_2.1.1              </span></span>
<span><span class="co">#&gt;  [47] DelayedMatrixStats_1.26.0   leiden_0.4.3.1             </span></span>
<span><span class="co">#&gt;  [49] codetools_0.2-20            DelayedArray_0.30.1        </span></span>
<span><span class="co">#&gt;  [51] scuttle_1.14.0              tidyselect_1.2.1           </span></span>
<span><span class="co">#&gt;  [53] UCSC.utils_1.0.0            farver_2.1.2               </span></span>
<span><span class="co">#&gt;  [55] viridis_0.6.5               ScaledMatrix_1.12.0        </span></span>
<span><span class="co">#&gt;  [57] matrixStats_1.4.1           stats4_4.4.1               </span></span>
<span><span class="co">#&gt;  [59] spatstat.explore_3.3-2      jsonlite_1.8.9             </span></span>
<span><span class="co">#&gt;  [61] BiocNeighbors_1.22.0        Formula_1.2-5              </span></span>
<span><span class="co">#&gt;  [63] progressr_0.14.0            ggridges_0.5.6             </span></span>
<span><span class="co">#&gt;  [65] survival_3.6-4              scater_1.32.1              </span></span>
<span><span class="co">#&gt;  [67] systemfonts_1.1.0           tools_4.4.1                </span></span>
<span><span class="co">#&gt;  [69] ragg_1.3.3                  ica_1.0-3                  </span></span>
<span><span class="co">#&gt;  [71] Rcpp_1.0.13                 glue_1.7.0                 </span></span>
<span><span class="co">#&gt;  [73] gridExtra_2.3               SparseArray_1.4.8          </span></span>
<span><span class="co">#&gt;  [75] xfun_0.47                   MatrixGenerics_1.16.0      </span></span>
<span><span class="co">#&gt;  [77] ggthemes_5.1.0              GenomeInfoDb_1.40.1        </span></span>
<span><span class="co">#&gt;  [79] withr_3.0.1                 formatR_1.14               </span></span>
<span><span class="co">#&gt;  [81] fastmap_1.2.0               fansi_1.0.6                </span></span>
<span><span class="co">#&gt;  [83] digest_0.6.37               rsvd_1.0.5                 </span></span>
<span><span class="co">#&gt;  [85] R6_2.5.1                    mime_0.12                  </span></span>
<span><span class="co">#&gt;  [87] textshaping_0.4.0           colorspace_2.1-1           </span></span>
<span><span class="co">#&gt;  [89] scattermore_1.2             tensor_1.5                 </span></span>
<span><span class="co">#&gt;  [91] spatstat.data_3.1-2         RhpcBLASctl_0.23-42        </span></span>
<span><span class="co">#&gt;  [93] utf8_1.2.4                  tidyr_1.3.1                </span></span>
<span><span class="co">#&gt;  [95] generics_0.1.3              data.table_1.16.0          </span></span>
<span><span class="co">#&gt;  [97] httr_1.4.7                  htmlwidgets_1.6.4          </span></span>
<span><span class="co">#&gt;  [99] S4Arrays_1.4.1              uwot_0.2.2                 </span></span>
<span><span class="co">#&gt; [101] pkgconfig_2.0.3             gtable_0.3.5               </span></span>
<span><span class="co">#&gt; [103] lmtest_0.9-40               SingleCellExperiment_1.26.0</span></span>
<span><span class="co">#&gt; [105] XVector_0.44.0              htmltools_0.5.8.1          </span></span>
<span><span class="co">#&gt; [107] carData_3.0-5               dotCall64_1.1-1            </span></span>
<span><span class="co">#&gt; [109] scales_1.3.0                Biobase_2.64.0             </span></span>
<span><span class="co">#&gt; [111] png_0.1-8                   harmony_1.2.1              </span></span>
<span><span class="co">#&gt; [113] spatstat.univar_3.0-1       knitr_1.48                 </span></span>
<span><span class="co">#&gt; [115] rstudioapi_0.16.0           reshape2_1.4.4             </span></span>
<span><span class="co">#&gt; [117] nlme_3.1-164                cachem_1.1.0               </span></span>
<span><span class="co">#&gt; [119] zoo_1.8-12                  stringr_1.5.1              </span></span>
<span><span class="co">#&gt; [121] KernSmooth_2.23-24          vipor_0.4.7                </span></span>
<span><span class="co">#&gt; [123] miniUI_0.1.1.1              GiRaF_1.0.1                </span></span>
<span><span class="co">#&gt; [125] desc_1.4.3                  pillar_1.9.0               </span></span>
<span><span class="co">#&gt; [127] grid_4.4.1                  vctrs_0.6.5                </span></span>
<span><span class="co">#&gt; [129] RANN_2.6.2                  ggpubr_0.6.0               </span></span>
<span><span class="co">#&gt; [131] promises_1.3.0              car_3.1-3                  </span></span>
<span><span class="co">#&gt; [133] BiocSingular_1.20.0         DR.SC_3.4                  </span></span>
<span><span class="co">#&gt; [135] beachmat_2.20.0             xtable_1.8-4               </span></span>
<span><span class="co">#&gt; [137] cluster_2.1.6               beeswarm_0.4.0             </span></span>
<span><span class="co">#&gt; [139] evaluate_1.0.0              cli_3.6.3                  </span></span>
<span><span class="co">#&gt; [141] compiler_4.4.1              rlang_1.1.4                </span></span>
<span><span class="co">#&gt; [143] crayon_1.5.3                ggsignif_0.6.4             </span></span>
<span><span class="co">#&gt; [145] future.apply_1.11.2         labeling_0.4.3             </span></span>
<span><span class="co">#&gt; [147] mclust_6.1.1                plyr_1.8.9                 </span></span>
<span><span class="co">#&gt; [149] fs_1.6.4                    ggbeeswarm_0.7.2           </span></span>
<span><span class="co">#&gt; [151] stringi_1.8.4               viridisLite_0.4.2          </span></span>
<span><span class="co">#&gt; [153] deldir_2.0-4                BiocParallel_1.38.0        </span></span>
<span><span class="co">#&gt; [155] munsell_0.5.1               lazyeval_0.2.2             </span></span>
<span><span class="co">#&gt; [157] spatstat.geom_3.3-3         CompQuadForm_1.4.3         </span></span>
<span><span class="co">#&gt; [159] RcppHNSW_0.6.0              patchwork_1.3.0            </span></span>
<span><span class="co">#&gt; [161] sparseMatrixStats_1.16.0    future_1.34.0              </span></span>
<span><span class="co">#&gt; [163] statmod_1.5.0               shiny_1.9.1                </span></span>
<span><span class="co">#&gt; [165] highr_0.11                  SummarizedExperiment_1.34.0</span></span>
<span><span class="co">#&gt; [167] ROCR_1.0-11                 broom_1.0.7                </span></span>
<span><span class="co">#&gt; [169] igraph_2.0.3                bslib_0.8.0</span></span></code></pre></div>
</details>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Wei Liu, Yi Yang, Jin Liu.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

      </footer>
</div>






  </body>
</html>
