<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>PRECAST: Four DLPFC Sample Analysis • PRECAST</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="PRECAST: Four DLPFC Sample Analysis">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">


    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">PRECAST</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.7</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/PRECAST.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/PRECAST.BreastCancer.html">PRECAST: Human Breast Cancer Data Analysis</a>
    </li>
    <li>
      <a href="../articles/PRECAST.DLPFC.html">PRECAST: DLPFC Single Sample Analysis</a>
    </li>
    <li>
      <a href="../articles/PRECAST.DLPFC4.html">PRECAST: Four DLPFC Sample Analysis</a>
    </li>
    <li>
      <a href="../articles/PRECAST.Simu.html">PRECAST: simulation</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/feiyoung/PRECAST/" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->



      </header><link href="PRECAST.DLPFC4_files/htmltools-fill-0.5.8.1/fill.css" rel="stylesheet">
<script src="PRECAST.DLPFC4_files/htmlwidgets-1.6.4/htmlwidgets.js"></script><link href="PRECAST.DLPFC4_files/datatables-css-0.0.0/datatables-crosstalk.css" rel="stylesheet">
<script src="PRECAST.DLPFC4_files/datatables-binding-0.33/datatables.js"></script><link href="PRECAST.DLPFC4_files/dt-core-1.13.6/css/jquery.dataTables.min.css" rel="stylesheet">
<link href="PRECAST.DLPFC4_files/dt-core-1.13.6/css/jquery.dataTables.extra.css" rel="stylesheet">
<script src="PRECAST.DLPFC4_files/dt-core-1.13.6/js/jquery.dataTables.min.js"></script><link href="PRECAST.DLPFC4_files/crosstalk-1.2.1/css/crosstalk.min.css" rel="stylesheet">
<script src="PRECAST.DLPFC4_files/crosstalk-1.2.1/js/crosstalk.min.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>PRECAST: Four DLPFC Sample Analysis</h1>
                        <h4 data-toc-skip class="author">Wei Liu</h4>
            
            <h4 data-toc-skip class="date">2025-09-29</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/feiyoung/PRECAST/blob/HEAD/vignettes/PRECAST.DLPFC4.Rmd" class="external-link"><code>vignettes/PRECAST.DLPFC4.Rmd</code></a></small>
      <div class="hidden name"><code>PRECAST.DLPFC4.Rmd</code></div>

    </div>

    
    
<p>This vignette introduces the PRECAST workflow for the analysis of
four spatial transcriptomics datasets. The workflow consists of three
steps</p>
<ul>
<li>Independent preprocessing and model setting</li>
<li>Probabilistic embedding and clustering using PRECAST model</li>
<li>Downstream analysis (i.e. visualization of clusters and
embeddings)</li>
</ul>
<p>We demonstrate the use of PRECAST to four human dorsolateral
prefrontal cortex Visium data. Download the data to R</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">SingleCellExperiment</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">name_ID4</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">151673</span>, <span class="fl">151674</span>, <span class="fl">151675</span>, <span class="fl">151676</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co">### Read data in an online manner</span></span>
<span><span class="va">n_ID</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">name_ID4</span><span class="op">)</span></span>
<span><span class="va">url_brainA</span> <span class="op">&lt;-</span> <span class="st">"https://github.com/feiyoung/DR-SC.Analysis/raw/main/data/DLPFC_data/"</span></span>
<span><span class="va">url_brainB</span> <span class="op">&lt;-</span> <span class="st">".rds"</span></span>
<span><span class="va">seuList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="va"><a href="https://github.com/feiyoung/ProFAST" class="external-link">ProFAST</a></span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="fu">remotes</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"feiyoung/ProFAST"</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="va">n_ID</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="co"># i &lt;- 1</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/cat.html" class="external-link">cat</a></span><span class="op">(</span><span class="st">"input brain data"</span>, <span class="va">i</span>, <span class="st">"\n"</span><span class="op">)</span></span>
<span>    <span class="co"># load and read data</span></span>
<span>    <span class="va">dlpfc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/connections.html" class="external-link">url</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">url_brainA</span>, <span class="va">name_ID4</span><span class="op">[</span><span class="va">i</span><span class="op">]</span>, <span class="va">url_brainB</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="co"># dlpfc &lt;- readRDS((paste0('', name_ID4[i],url_brainB) ))</span></span>
<span>    <span class="va">count</span> <span class="op">&lt;-</span> <span class="va">dlpfc</span><span class="op">@</span><span class="va">assays</span><span class="op">@</span><span class="va">data</span><span class="op">$</span><span class="va">counts</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/row.names.html" class="external-link">row.names</a></span><span class="op">(</span><span class="va">count</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu">ProFAST</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ProFAST/man/transferGeneNames.html" class="external-link">transferGeneNames</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/row.names.html" class="external-link">row.names</a></span><span class="op">(</span><span class="va">count</span><span class="op">)</span>, species <span class="op">=</span> <span class="st">"Human"</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/row.names.html" class="external-link">row.names</a></span><span class="op">(</span><span class="va">count</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/make.unique.html" class="external-link">make.unique</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/row.names.html" class="external-link">row.names</a></span><span class="op">(</span><span class="va">count</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="va">seu1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.github.io/seurat-object/reference/CreateSeuratObject.html" class="external-link">CreateSeuratObject</a></span><span class="op">(</span>counts <span class="op">=</span> <span class="va">count</span>, meta.data <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">dlpfc</span><span class="op">)</span><span class="op">)</span>, min.cells <span class="op">=</span> <span class="fl">10</span>,</span>
<span>        min.features <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span>
<span>    <span class="va">seuList</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">seu1</span></span>
<span><span class="op">}</span></span>
<span><span class="co"># saveRDS(seuList, file='seuList4.RDS')</span></span></code></pre></div>
<p>The package can be loaded with the command:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/feiyoung/PRECAST" class="external-link">PRECAST</a></span><span class="op">)</span></span></code></pre></div>
<div class="section level2">
<h2 id="compare-precast-with-dr-sc-in-analyzing-one-sample">Compare PRECAST with DR-SC in analyzing one sample<a class="anchor" aria-label="anchor" href="#compare-precast-with-dr-sc-in-analyzing-one-sample"></a>
</h2>
<p>First, we view the the spatial transcriptomics data with Visium
platform.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seuList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">"seuList4.RDS"</span><span class="op">)</span></span>
<span><span class="va">seuList</span>  <span class="co">## a list including  Seurat objects</span></span>
<span><span class="co">#&gt; [[1]]</span></span>
<span><span class="co">#&gt; An object of class Seurat </span></span>
<span><span class="co">#&gt; 16578 features across 3639 samples within 1 assay </span></span>
<span><span class="co">#&gt; Active assay: RNA (16578 features, 0 variable features)</span></span>
<span><span class="co">#&gt;  1 layer present: counts</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; [[2]]</span></span>
<span><span class="co">#&gt; An object of class Seurat </span></span>
<span><span class="co">#&gt; 17332 features across 3673 samples within 1 assay </span></span>
<span><span class="co">#&gt; Active assay: RNA (17332 features, 0 variable features)</span></span>
<span><span class="co">#&gt;  1 layer present: counts</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; [[3]]</span></span>
<span><span class="co">#&gt; An object of class Seurat </span></span>
<span><span class="co">#&gt; 16062 features across 3592 samples within 1 assay </span></span>
<span><span class="co">#&gt; Active assay: RNA (16062 features, 0 variable features)</span></span>
<span><span class="co">#&gt;  1 layer present: counts</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; [[4]]</span></span>
<span><span class="co">#&gt; An object of class Seurat </span></span>
<span><span class="co">#&gt; 16097 features across 3460 samples within 1 assay </span></span>
<span><span class="co">#&gt; Active assay: RNA (16097 features, 0 variable features)</span></span>
<span><span class="co">#&gt;  1 layer present: counts</span></span></code></pre></div>
<p>check the meta data that must include the spatial coordinates named
“row” and “col”, respectively. If the names are not, they are required
to rename them.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">metadataList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">seuList</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="va">x</span><span class="op">@</span><span class="va">meta.data</span><span class="op">)</span></span>
<span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">r</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_along</a></span><span class="op">(</span><span class="va">metadataList</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">meta_data</span> <span class="op">&lt;-</span> <span class="va">metadataList</span><span class="op">[[</span><span class="va">r</span><span class="op">]</span><span class="op">]</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/cat.html" class="external-link">cat</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/all.html" class="external-link">all</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"row"</span>, <span class="st">"col"</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">meta_data</span><span class="op">)</span><span class="op">)</span>, <span class="st">"\n"</span><span class="op">)</span>  <span class="co">## the names are correct!</span></span>
<span></span>
<span><span class="op">}</span></span>
<span><span class="co">#&gt; TRUE </span></span>
<span><span class="co">#&gt; TRUE </span></span>
<span><span class="co">#&gt; TRUE </span></span>
<span><span class="co">#&gt; TRUE</span></span></code></pre></div>
<div class="section level3">
<h3 id="prepare-the-precastobject-">Prepare the PRECASTObject.<a class="anchor" aria-label="anchor" href="#prepare-the-precastobject-"></a>
</h3>
<p>Create a PRECASTObj object to prepare for PRECAST models. Here, we
only show the <code>HVGs</code> method to select the 2000 highly
variable genes, but users are able to choose more genes or also use the
<code>SPARK-X</code> to choose the spatially variable genes.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">2023</span><span class="op">)</span></span>
<span><span class="va">preobj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/CreateiDRSCObject.html">CreatePRECASTObject</a></span><span class="op">(</span>seuList <span class="op">=</span> <span class="va">seuList</span>, selectGenesMethod <span class="op">=</span> <span class="st">"HVGs"</span>, gene.number <span class="op">=</span> <span class="fl">2000</span><span class="op">)</span>  <span class="co"># </span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="add-the-model-setting">Add the model setting<a class="anchor" aria-label="anchor" href="#add-the-model-setting"></a>
</h3>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## check the number of genes/features after filtering step</span></span>
<span><span class="va">preobj</span><span class="op">@</span><span class="va">seulist</span></span>
<span><span class="co">#&gt; [[1]]</span></span>
<span><span class="co">#&gt; An object of class Seurat </span></span>
<span><span class="co">#&gt; 2000 features across 3639 samples within 1 assay </span></span>
<span><span class="co">#&gt; Active assay: RNA (2000 features, 1304 variable features)</span></span>
<span><span class="co">#&gt;  2 layers present: counts, data</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; [[2]]</span></span>
<span><span class="co">#&gt; An object of class Seurat </span></span>
<span><span class="co">#&gt; 2000 features across 3673 samples within 1 assay </span></span>
<span><span class="co">#&gt; Active assay: RNA (2000 features, 1296 variable features)</span></span>
<span><span class="co">#&gt;  2 layers present: counts, data</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; [[3]]</span></span>
<span><span class="co">#&gt; An object of class Seurat </span></span>
<span><span class="co">#&gt; 2000 features across 3592 samples within 1 assay </span></span>
<span><span class="co">#&gt; Active assay: RNA (2000 features, 1363 variable features)</span></span>
<span><span class="co">#&gt;  2 layers present: counts, data</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; [[4]]</span></span>
<span><span class="co">#&gt; An object of class Seurat </span></span>
<span><span class="co">#&gt; 2000 features across 3460 samples within 1 assay </span></span>
<span><span class="co">#&gt; Active assay: RNA (2000 features, 1311 variable features)</span></span>
<span><span class="co">#&gt;  2 layers present: counts, data</span></span>
<span><span class="co">## Add adjacency matrix list for a PRECASTObj object to prepare for PRECAST model fitting.</span></span>
<span><span class="va">PRECASTObj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/AddAdjList.html">AddAdjList</a></span><span class="op">(</span><span class="va">preobj</span>, platform <span class="op">=</span> <span class="st">"Visium"</span><span class="op">)</span></span>
<span><span class="co">## Add a model setting in advance for a PRECASTObj object. verbose =TRUE helps outputing the</span></span>
<span><span class="co">## information in the algorithm. We provide two initial model to obtain initial values:</span></span>
<span><span class="co">## 'mclust' and 'kmeans', and recommend using 'mclust' in LIBD data.</span></span>
<span><span class="va">PRECASTObj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/AddParSetting.html">AddParSetting</a></span><span class="op">(</span><span class="va">PRECASTObj</span>, Sigma_equal <span class="op">=</span> <span class="cn">TRUE</span>, coreNum <span class="op">=</span> <span class="fl">1</span>, int.model <span class="op">=</span> <span class="st">"mclust"</span>, maxIter <span class="op">=</span> <span class="fl">30</span>,</span>
<span>    verbose <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="fit-precast">Fit PRECAST<a class="anchor" aria-label="anchor" href="#fit-precast"></a>
</h3>
<p>For function <code>PRECAST</code>, users can specify the number of
clusters
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>K</mi><annotation encoding="application/x-tex">K</annotation></semantics></math>
or set <code>K</code> to be an integer vector by using modified
BIC(MBIC) to determine
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>K</mi><annotation encoding="application/x-tex">K</annotation></semantics></math>.
Here, we use user-specified number of clusters.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">### Given K</span></span>
<span><span class="va">PRECASTObj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/PRECAST.html">PRECAST</a></span><span class="op">(</span><span class="va">PRECASTObj</span>, K <span class="op">=</span> <span class="fl">7</span><span class="op">)</span></span>
<span><span class="co">#&gt; iter = 2, loglik= -1751084.625000, dloglik=0.999185 </span></span>
<span><span class="co">#&gt; iter = 3, loglik= -1687448.000000, dloglik=0.036341 </span></span>
<span><span class="co">#&gt; iter = 4, loglik= -1663030.375000, dloglik=0.014470 </span></span>
<span><span class="co">#&gt; iter = 5, loglik= -1651564.625000, dloglik=0.006894 </span></span>
<span><span class="co">#&gt; iter = 6, loglik= -1645268.625000, dloglik=0.003812 </span></span>
<span><span class="co">#&gt; iter = 7, loglik= -1641603.750000, dloglik=0.002228 </span></span>
<span><span class="co">#&gt; iter = 8, loglik= -1639368.875000, dloglik=0.001361 </span></span>
<span><span class="co">#&gt; iter = 9, loglik= -1637948.500000, dloglik=0.000866 </span></span>
<span><span class="co">#&gt; iter = 10, loglik= -1636986.750000, dloglik=0.000587 </span></span>
<span><span class="co">#&gt; iter = 11, loglik= -1636350.375000, dloglik=0.000389 </span></span>
<span><span class="co">#&gt; iter = 12, loglik= -1635821.500000, dloglik=0.000323 </span></span>
<span><span class="co">#&gt; iter = 13, loglik= -1635491.625000, dloglik=0.000202 </span></span>
<span><span class="co">#&gt; iter = 14, loglik= -1635213.500000, dloglik=0.000170 </span></span>
<span><span class="co">#&gt; iter = 15, loglik= -1635032.125000, dloglik=0.000111 </span></span>
<span><span class="co">#&gt; iter = 16, loglik= -1634857.625000, dloglik=0.000107 </span></span>
<span><span class="co">#&gt; iter = 17, loglik= -1634731.500000, dloglik=0.000077 </span></span>
<span><span class="co">#&gt; iter = 18, loglik= -1634608.750000, dloglik=0.000075 </span></span>
<span><span class="co">#&gt; iter = 19, loglik= -1634541.125000, dloglik=0.000041 </span></span>
<span><span class="co">#&gt; iter = 20, loglik= -1634495.875000, dloglik=0.000028 </span></span>
<span><span class="co">#&gt; iter = 21, loglik= -1634410.625000, dloglik=0.000052 </span></span>
<span><span class="co">#&gt; iter = 22, loglik= -1634362.750000, dloglik=0.000029 </span></span>
<span><span class="co">#&gt; iter = 23, loglik= -1634362.125000, dloglik=0.000000</span></span></code></pre></div>
<p>Use the function <code><a href="../reference/selectModel.html">SelectModel()</a></code> to re-organize the fitted
results in PRECASTObj.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## backup the fitting results in resList</span></span>
<span><span class="va">resList</span> <span class="op">&lt;-</span> <span class="va">PRECASTObj</span><span class="op">@</span><span class="va">resList</span></span>
<span><span class="va">PRECASTObj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/selectModel.html">SelectModel</a></span><span class="op">(</span><span class="va">PRECASTObj</span><span class="op">)</span></span>
<span><span class="va">ari_precast</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">seuList</span><span class="op">)</span>, <span class="kw">function</span><span class="op">(</span><span class="va">r</span><span class="op">)</span> <span class="fu">mclust</span><span class="fu">::</span><span class="fu"><a href="https://mclust-org.github.io/mclust/reference/adjustedRandIndex.html" class="external-link">adjustedRandIndex</a></span><span class="op">(</span><span class="va">PRECASTObj</span><span class="op">@</span><span class="va">resList</span><span class="op">$</span><span class="va">cluster</span><span class="op">[[</span><span class="va">r</span><span class="op">]</span><span class="op">]</span>,</span>
<span>    <span class="va">PRECASTObj</span><span class="op">@</span><span class="va">seulist</span><span class="op">[[</span><span class="va">r</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">layer_guess_reordered</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">mat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">round</a></span><span class="op">(</span><span class="va">ari_precast</span>, <span class="fl">2</span><span class="op">)</span>, nrow <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">name_ID4</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">151673</span>, <span class="fl">151674</span>, <span class="fl">151675</span>, <span class="fl">151676</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">mat</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">name_ID4</span></span>
<span><span class="fu">DT</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/DT/man/datatable.html" class="external-link">datatable</a></span><span class="op">(</span><span class="va">mat</span><span class="op">)</span></span></code></pre></div>
<div class="datatables html-widget html-fill-item" id="htmlwidget-ac96cb3ee4656e2e9ec3" style="width:100%;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-ac96cb3ee4656e2e9ec3">{"x":{"filter":"none","vertical":false,"data":[[0.5600000000000001],[0.6],[0.5600000000000001],[0.57]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th>151673<\/th>\n      <th>151674<\/th>\n      <th>151675<\/th>\n      <th>151676<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"columnDefs":[{"className":"dt-right","targets":[0,1,2,3]},{"name":"151673","targets":0},{"name":"151674","targets":1},{"name":"151675","targets":2},{"name":"151676","targets":3}],"order":[],"autoWidth":false,"orderClasses":false},"selection":{"mode":"multiple","selected":null,"target":"row","selectable":null}},"evals":[],"jsHooks":[]}</script><details><summary><strong>Other options</strong>
</summary><p>Users are also able to set multiple K, then choose the best one</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">PRECASTObj2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/AddParSetting.html">AddParSetting</a></span><span class="op">(</span><span class="va">PRECASTObj</span>, Sigma_equal <span class="op">=</span> <span class="cn">FALSE</span>, coreNum <span class="op">=</span> <span class="fl">4</span>, maxIter <span class="op">=</span> <span class="fl">30</span>, verbose <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>  <span class="co"># set 4 cores to run in parallel.</span></span>
<span><span class="va">PRECASTObj2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/PRECAST.html">PRECAST</a></span><span class="op">(</span><span class="va">PRECASTObj2</span>, K <span class="op">=</span> <span class="fl">5</span><span class="op">:</span><span class="fl">8</span><span class="op">)</span></span>
<span><span class="co">## backup the fitting results in resList</span></span>
<span><span class="va">resList2</span> <span class="op">&lt;-</span> <span class="va">PRECASTObj2</span><span class="op">@</span><span class="va">resList</span></span>
<span><span class="va">PRECASTObj2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/selectModel.html">SelectModel</a></span><span class="op">(</span><span class="va">PRECASTObj2</span><span class="op">)</span></span></code></pre></div>
<p>Besides, user can also use different initialization method by setting
<code>int.model</code>, for example, set <code>int.model=NULL</code>;
see the functions <code><a href="../reference/AddParSetting.html">AddParSetting()</a></code> and
<code><a href="../reference/model_set.html">model_set()</a></code> for more details.</p>
</details>
</div>
<div class="section level3">
<h3 id="integration-and-visualization">Integration and Visualization<a class="anchor" aria-label="anchor" href="#integration-and-visualization"></a>
</h3>
<p>Integrate the all samples using the <code>IntegrateSpaData</code>
function. For computational efficiency, this function exclusively
integrates the variable genes. Specifically, in cases where users do not
specify the <code>PRECASTObj@seuList</code> or <code>seuList</code>
argument within the <code>IntegrateSpaData</code> function, it
automatically focuses on integrating only the variable genes. The
default setting for <code>PRECASTObj@seuList</code> is <code>NULL</code>
when <code>rawData.preserve</code> in <code>CreatePRECASTObject</code>
is set to <code>FALSE</code>. For instance:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">PRECASTObj</span><span class="op">@</span><span class="va">seuList</span><span class="op">)</span></span>
<span><span class="co">#&gt; NULL</span></span>
<span><span class="va">seuInt</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/IntegrateSpaData.html">IntegrateSpaData</a></span><span class="op">(</span><span class="va">PRECASTObj</span>, species <span class="op">=</span> <span class="st">"Human"</span><span class="op">)</span></span>
<span><span class="va">seuInt</span></span>
<span><span class="co">#&gt; An object of class Seurat </span></span>
<span><span class="co">#&gt; 2000 features across 14364 samples within 1 assay </span></span>
<span><span class="co">#&gt; Active assay: PRE_CAST (2000 features, 0 variable features)</span></span>
<span><span class="co">#&gt;  2 layers present: counts, data</span></span>
<span><span class="co">#&gt;  2 dimensional reductions calculated: PRECAST, position</span></span>
<span><span class="co">## The low-dimensional embeddings obtained by PRECAST are saved in PRECAST reduction slot.</span></span></code></pre></div>
<details><summary><strong>Integrating all genes</strong>
</summary><p>There are two ways to use <code>IntegrateSpaData</code> integrating
all genes, which will require more memory. We recommand running for all
genes on server. The first one is to set value for
<code>PRECASTObj@seuList</code>.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## assign the raw Seurat list object to it.  For illustration, we generate a new seuList with</span></span>
<span><span class="co">## more genes; For integrating all genes, users can set `seuList &lt;- bc2`.</span></span>
<span><span class="va">genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/row.names.html" class="external-link">row.names</a></span><span class="op">(</span><span class="va">PRECASTObj</span><span class="op">@</span><span class="va">seulist</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/row.names.html" class="external-link">row.names</a></span><span class="op">(</span><span class="va">seuList</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="va">seuList_sub</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">seuList</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="va">x</span><span class="op">[</span><span class="va">genes</span>, <span class="op">]</span><span class="op">)</span></span>
<span><span class="va">PRECASTObj</span><span class="op">@</span><span class="va">seuList</span> <span class="op">&lt;-</span> <span class="va">seuList_sub</span>  <span class="co"># </span></span>
<span><span class="va">seuInt</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/IntegrateSpaData.html">IntegrateSpaData</a></span><span class="op">(</span><span class="va">PRECASTObj</span>, species <span class="op">=</span> <span class="st">"Human"</span><span class="op">)</span></span>
<span><span class="va">seuInt</span></span></code></pre></div>
<p>The second method is to set a value for the argument
<code>seuList</code>:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">PRECASTObj</span><span class="op">@</span><span class="va">seuList</span> <span class="op">&lt;-</span> <span class="cn">NULL</span></span>
<span><span class="co">## At the same time, we can set subsampling to speed up the computation.</span></span>
<span><span class="va">seuInt</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/IntegrateSpaData.html">IntegrateSpaData</a></span><span class="op">(</span><span class="va">PRECASTObj</span>, species <span class="op">=</span> <span class="st">"Human"</span>, seuList <span class="op">=</span> <span class="va">seuList_sub</span>, subsample_rate <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span></span>
<span><span class="va">seuInt</span></span></code></pre></div>
</details><p>First, user can choose a beautiful color schema using
<code><a href="../reference/chooseColors.html">chooseColors()</a></code>.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cols_cluster</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/chooseColors.html">chooseColors</a></span><span class="op">(</span>palettes_name <span class="op">=</span> <span class="st">"Classic 20"</span>, n_colors <span class="op">=</span> <span class="fl">7</span>, plot_colors <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="PRECAST.DLPFC4_files/figure-html/unnamed-chunk-14-1.png" width="1056"></p>
<p>Show the spatial scatter plot for clusters</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">p12</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/SpaPlot.html">SpaPlot</a></span><span class="op">(</span><span class="va">seuInt</span>, item <span class="op">=</span> <span class="st">"cluster"</span>, batch <span class="op">=</span> <span class="cn">NULL</span>, point_size <span class="op">=</span> <span class="fl">1</span>, cols <span class="op">=</span> <span class="va">cols_cluster</span>, combine <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    nrow.legend <span class="op">=</span> <span class="fl">7</span><span class="op">)</span></span>
<span><span class="va">p12</span></span></code></pre></div>
<p><img src="PRECAST.DLPFC4_files/figure-html/unnamed-chunk-15-1.png" width="864"></p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># users can plot each sample by setting combine=FALSE</span></span></code></pre></div>
<p>Users can re-plot the above figures for specific need by returning a
ggplot list object. For example, we plot the spatial heatmap using a
common legend.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="va">pList</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/SpaPlot.html">SpaPlot</a></span><span class="op">(</span><span class="va">seuInt</span>, item <span class="op">=</span> <span class="st">"cluster"</span>, batch <span class="op">=</span> <span class="cn">NULL</span>, point_size <span class="op">=</span> <span class="fl">2.5</span>, cols <span class="op">=</span> <span class="va">cols_cluster</span>,</span>
<span>    combine <span class="op">=</span> <span class="cn">FALSE</span>, nrow.legend <span class="op">=</span> <span class="fl">7</span><span class="op">)</span></span>
<span><span class="va">pList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">pList</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="va">x</span> <span class="op">+</span> <span class="fu"><a href="https://rdrr.io/pkg/ggplot2/man/coord_flip.html" class="external-link">coord_flip</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://rdrr.io/pkg/ggplot2/man/scale_continuous.html" class="external-link">scale_x_reverse</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/drawFigs.html">drawFigs</a></span><span class="op">(</span><span class="va">pList</span>, layout.dim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2</span>, <span class="fl">2</span><span class="op">)</span>, common.legend <span class="op">=</span> <span class="cn">TRUE</span>, legend.position <span class="op">=</span> <span class="st">"right"</span>, align <span class="op">=</span> <span class="st">"hv"</span><span class="op">)</span></span></code></pre></div>
<p><img src="PRECAST.DLPFC4_files/figure-html/unnamed-chunk-16-1.png" width="816"></p>
<p>Show the spatial UMAP/tNSE RGB plot to illustrate the performance in
extracting features.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seuInt</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/AddUMAP.html">AddUMAP</a></span><span class="op">(</span><span class="va">seuInt</span><span class="op">)</span></span>
<span><span class="va">p13List</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/SpaPlot.html">SpaPlot</a></span><span class="op">(</span><span class="va">seuInt</span>, batch <span class="op">=</span> <span class="cn">NULL</span>, item <span class="op">=</span> <span class="st">"RGB_UMAP"</span>, point_size <span class="op">=</span> <span class="fl">2</span>, combine <span class="op">=</span> <span class="cn">FALSE</span>, text_size <span class="op">=</span> <span class="fl">15</span><span class="op">)</span></span>
<span><span class="va">p13List</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">p13List</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="va">x</span> <span class="op">+</span> <span class="fu"><a href="https://rdrr.io/pkg/ggplot2/man/coord_flip.html" class="external-link">coord_flip</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://rdrr.io/pkg/ggplot2/man/scale_continuous.html" class="external-link">scale_x_reverse</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/drawFigs.html">drawFigs</a></span><span class="op">(</span><span class="va">p13List</span>, layout.dim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2</span>, <span class="fl">2</span><span class="op">)</span>, common.legend <span class="op">=</span> <span class="cn">TRUE</span>, legend.position <span class="op">=</span> <span class="st">"right"</span>, align <span class="op">=</span> <span class="st">"hv"</span><span class="op">)</span></span></code></pre></div>
<p><img src="PRECAST.DLPFC4_files/figure-html/unnamed-chunk-17-1.png" width="528"></p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># seuInt &lt;- AddTSNE(seuInt) SpaPlot(seuInt, batch=NULL,item='RGB_TSNE',point_size=2,</span></span>
<span><span class="co"># combine=T, text_size=15)</span></span></code></pre></div>
<p>Show the tSNE plot based on the extracted features from PRECAST to
check the performance of integration.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seuInt</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/AddTSNE.html">AddTSNE</a></span><span class="op">(</span><span class="va">seuInt</span>, n_comp <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span>
<span><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/dimPlot.html">dimPlot</a></span><span class="op">(</span><span class="va">seuInt</span>, item <span class="op">=</span> <span class="st">"cluster"</span>, point_size <span class="op">=</span> <span class="fl">0.5</span>, font_family <span class="op">=</span> <span class="st">"serif"</span>, cols <span class="op">=</span> <span class="va">cols_cluster</span>,</span>
<span>    border_col <span class="op">=</span> <span class="st">"gray10"</span>, nrow.legend <span class="op">=</span> <span class="fl">14</span>, legend_pos <span class="op">=</span> <span class="st">"right"</span><span class="op">)</span>  <span class="co"># Times New Roman</span></span>
<span><span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/dimPlot.html">dimPlot</a></span><span class="op">(</span><span class="va">seuInt</span>, item <span class="op">=</span> <span class="st">"batch"</span>, point_size <span class="op">=</span> <span class="fl">0.5</span>, font_family <span class="op">=</span> <span class="st">"serif"</span>, legend_pos <span class="op">=</span> <span class="st">"right"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/drawFigs.html">drawFigs</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">p1</span>, <span class="va">p2</span><span class="op">)</span>, layout.dim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span><span class="op">)</span>, legend.position <span class="op">=</span> <span class="st">"right"</span>, align <span class="op">=</span> <span class="st">"hv"</span><span class="op">)</span></span></code></pre></div>
<p><img src="PRECAST.DLPFC4_files/figure-html/unnamed-chunk-18-1.png" width="1152"></p>
<p>Combined differential expression analysis</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span></span>
<span><span class="va">dat_deg</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindAllMarkers.html" class="external-link">FindAllMarkers</a></span><span class="op">(</span><span class="va">seuInt</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span><span class="va">n</span> <span class="op">&lt;-</span> <span class="fl">5</span></span>
<span><span class="va">dat_deg</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/dplyr/man/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">cluster</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/dplyr/man/top_n.html" class="external-link">top_n</a></span><span class="op">(</span>n <span class="op">=</span> <span class="va">n</span>, wt <span class="op">=</span> <span class="va">avg_log2FC</span><span class="op">)</span> <span class="op">-&gt;</span> <span class="va">top10</span></span></code></pre></div>
<p>Plot DE genes’ heatmap for each spatial domain identified by
PRECAST.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="co">## HeatMap</span></span>
<span><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/DotPlot.html" class="external-link">DotPlot</a></span><span class="op">(</span><span class="va">seuInt</span>, features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">top10</span><span class="op">$</span><span class="va">gene</span><span class="op">)</span>, col.min <span class="op">=</span> <span class="fl">0</span>, col.max <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://rdrr.io/pkg/ggplot2/man/theme.html" class="external-link">theme</a></span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/ggplot2/man/element.html" class="external-link">element_text</a></span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">45</span>,</span>
<span>    hjust <span class="op">=</span> <span class="fl">1</span>, size <span class="op">=</span> <span class="fl">8</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">p1</span></span></code></pre></div>
<p><img src="PRECAST.DLPFC4_files/figure-html/unnamed-chunk-20-1.png" width="768"></p>
<details><summary><strong>Session Info</strong>
</summary><div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.4.1 (2024-06-14 ucrt)</span></span>
<span><span class="co">#&gt; Platform: x86_64-w64-mingw32/x64</span></span>
<span><span class="co">#&gt; Running under: Windows 11 x64 (build 26100)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt; [1] LC_COLLATE=Chinese (Simplified)_China.utf8 </span></span>
<span><span class="co">#&gt; [2] LC_CTYPE=Chinese (Simplified)_China.utf8   </span></span>
<span><span class="co">#&gt; [3] LC_MONETARY=Chinese (Simplified)_China.utf8</span></span>
<span><span class="co">#&gt; [4] LC_NUMERIC=C                               </span></span>
<span><span class="co">#&gt; [5] LC_TIME=Chinese (Simplified)_China.utf8    </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: Asia/Shanghai</span></span>
<span><span class="co">#&gt; tzcode source: internal</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] parallel  stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">#&gt; [8] base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt; [1] dplyr_1.1.4        Seurat_5.1.0       SeuratObject_5.0.2 sp_2.1-4          </span></span>
<span><span class="co">#&gt; [5] ggplot2_3.5.2      irlba_2.3.5.1      Matrix_1.7-0       PRECAST_1.7       </span></span>
<span><span class="co">#&gt; [9] gtools_3.9.5      </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;   [1] RcppAnnoy_0.0.22            splines_4.4.1              </span></span>
<span><span class="co">#&gt;   [3] later_1.3.2                 tibble_3.2.1               </span></span>
<span><span class="co">#&gt;   [5] polyclip_1.10-7             fastDummies_1.7.4          </span></span>
<span><span class="co">#&gt;   [7] lifecycle_1.0.4             rstatix_0.7.2              </span></span>
<span><span class="co">#&gt;   [9] globals_0.16.3              lattice_0.22-6             </span></span>
<span><span class="co">#&gt;  [11] MASS_7.3-60.2               crosstalk_1.2.1            </span></span>
<span><span class="co">#&gt;  [13] backports_1.5.0             magrittr_2.0.3             </span></span>
<span><span class="co">#&gt;  [15] limma_3.58.1                plotly_4.10.4              </span></span>
<span><span class="co">#&gt;  [17] sass_0.4.9                  rmarkdown_2.28             </span></span>
<span><span class="co">#&gt;  [19] jquerylib_0.1.4             yaml_2.3.10                </span></span>
<span><span class="co">#&gt;  [21] httpuv_1.6.15               sctransform_0.4.1          </span></span>
<span><span class="co">#&gt;  [23] spam_2.10-0                 spatstat.sparse_3.1-0      </span></span>
<span><span class="co">#&gt;  [25] reticulate_1.39.0           cowplot_1.1.3              </span></span>
<span><span class="co">#&gt;  [27] pbapply_1.7-2               RColorBrewer_1.1-3         </span></span>
<span><span class="co">#&gt;  [29] abind_1.4-8                 zlibbioc_1.50.0            </span></span>
<span><span class="co">#&gt;  [31] Rtsne_0.17                  GenomicRanges_1.56.2       </span></span>
<span><span class="co">#&gt;  [33] presto_1.0.0                purrr_1.0.2                </span></span>
<span><span class="co">#&gt;  [35] BiocGenerics_0.50.0         GenomeInfoDbData_1.2.12    </span></span>
<span><span class="co">#&gt;  [37] IRanges_2.38.1              S4Vectors_0.42.1           </span></span>
<span><span class="co">#&gt;  [39] ggrepel_0.9.6               listenv_0.9.1              </span></span>
<span><span class="co">#&gt;  [41] spatstat.utils_3.1-0        goftest_1.2-3              </span></span>
<span><span class="co">#&gt;  [43] RSpectra_0.16-2             spatstat.random_3.3-2      </span></span>
<span><span class="co">#&gt;  [45] fitdistrplus_1.2-1          parallelly_1.38.0          </span></span>
<span><span class="co">#&gt;  [47] pkgdown_2.1.1               DelayedMatrixStats_1.26.0  </span></span>
<span><span class="co">#&gt;  [49] leiden_0.4.3.1              codetools_0.2-20           </span></span>
<span><span class="co">#&gt;  [51] DelayedArray_0.30.1         DT_0.33                    </span></span>
<span><span class="co">#&gt;  [53] scuttle_1.14.0              tidyselect_1.2.1           </span></span>
<span><span class="co">#&gt;  [55] UCSC.utils_1.0.0            farver_2.1.2               </span></span>
<span><span class="co">#&gt;  [57] viridis_0.6.5               ScaledMatrix_1.12.0        </span></span>
<span><span class="co">#&gt;  [59] matrixStats_1.4.1           stats4_4.4.1               </span></span>
<span><span class="co">#&gt;  [61] spatstat.explore_3.3-2      jsonlite_1.8.9             </span></span>
<span><span class="co">#&gt;  [63] BiocNeighbors_1.22.0        Formula_1.2-5              </span></span>
<span><span class="co">#&gt;  [65] progressr_0.14.0            ggridges_0.5.6             </span></span>
<span><span class="co">#&gt;  [67] survival_3.6-4              scater_1.32.1              </span></span>
<span><span class="co">#&gt;  [69] systemfonts_1.1.0           tools_4.4.1                </span></span>
<span><span class="co">#&gt;  [71] ragg_1.3.3                  ica_1.0-3                  </span></span>
<span><span class="co">#&gt;  [73] Rcpp_1.0.13                 glue_1.7.0                 </span></span>
<span><span class="co">#&gt;  [75] gridExtra_2.3               SparseArray_1.4.8          </span></span>
<span><span class="co">#&gt;  [77] xfun_0.47                   MatrixGenerics_1.16.0      </span></span>
<span><span class="co">#&gt;  [79] ggthemes_5.1.0              GenomeInfoDb_1.40.1        </span></span>
<span><span class="co">#&gt;  [81] withr_3.0.1                 formatR_1.14               </span></span>
<span><span class="co">#&gt;  [83] fastmap_1.2.0               fansi_1.0.6                </span></span>
<span><span class="co">#&gt;  [85] digest_0.6.37               rsvd_1.0.5                 </span></span>
<span><span class="co">#&gt;  [87] R6_2.5.1                    mime_0.12                  </span></span>
<span><span class="co">#&gt;  [89] textshaping_0.4.0           colorspace_2.1-1           </span></span>
<span><span class="co">#&gt;  [91] scattermore_1.2             tensor_1.5                 </span></span>
<span><span class="co">#&gt;  [93] spatstat.data_3.1-2         RhpcBLASctl_0.23-42        </span></span>
<span><span class="co">#&gt;  [95] utf8_1.2.4                  tidyr_1.3.1                </span></span>
<span><span class="co">#&gt;  [97] generics_0.1.3              data.table_1.16.0          </span></span>
<span><span class="co">#&gt;  [99] httr_1.4.7                  htmlwidgets_1.6.4          </span></span>
<span><span class="co">#&gt; [101] S4Arrays_1.4.1              uwot_0.2.2                 </span></span>
<span><span class="co">#&gt; [103] pkgconfig_2.0.3             gtable_0.3.5               </span></span>
<span><span class="co">#&gt; [105] lmtest_0.9-40               SingleCellExperiment_1.26.0</span></span>
<span><span class="co">#&gt; [107] XVector_0.44.0              htmltools_0.5.8.1          </span></span>
<span><span class="co">#&gt; [109] carData_3.0-5               dotCall64_1.1-1            </span></span>
<span><span class="co">#&gt; [111] scales_1.3.0                Biobase_2.64.0             </span></span>
<span><span class="co">#&gt; [113] png_0.1-8                   harmony_1.2.1              </span></span>
<span><span class="co">#&gt; [115] spatstat.univar_3.0-1       knitr_1.48                 </span></span>
<span><span class="co">#&gt; [117] rstudioapi_0.16.0           reshape2_1.4.4             </span></span>
<span><span class="co">#&gt; [119] nlme_3.1-164                cachem_1.1.0               </span></span>
<span><span class="co">#&gt; [121] zoo_1.8-12                  stringr_1.5.1              </span></span>
<span><span class="co">#&gt; [123] KernSmooth_2.23-24          vipor_0.4.7                </span></span>
<span><span class="co">#&gt; [125] miniUI_0.1.1.1              GiRaF_1.0.1                </span></span>
<span><span class="co">#&gt; [127] desc_1.4.3                  pillar_1.9.0               </span></span>
<span><span class="co">#&gt; [129] grid_4.4.1                  vctrs_0.6.5                </span></span>
<span><span class="co">#&gt; [131] RANN_2.6.2                  ggpubr_0.6.0               </span></span>
<span><span class="co">#&gt; [133] promises_1.3.0              car_3.1-3                  </span></span>
<span><span class="co">#&gt; [135] BiocSingular_1.20.0         DR.SC_3.4                  </span></span>
<span><span class="co">#&gt; [137] beachmat_2.20.0             xtable_1.8-4               </span></span>
<span><span class="co">#&gt; [139] cluster_2.1.6               beeswarm_0.4.0             </span></span>
<span><span class="co">#&gt; [141] evaluate_1.0.0              cli_3.6.3                  </span></span>
<span><span class="co">#&gt; [143] compiler_4.4.1              rlang_1.1.4                </span></span>
<span><span class="co">#&gt; [145] crayon_1.5.3                ggsignif_0.6.4             </span></span>
<span><span class="co">#&gt; [147] future.apply_1.11.2         labeling_0.4.3             </span></span>
<span><span class="co">#&gt; [149] mclust_6.1.1                plyr_1.8.9                 </span></span>
<span><span class="co">#&gt; [151] fs_1.6.4                    ggbeeswarm_0.7.2           </span></span>
<span><span class="co">#&gt; [153] stringi_1.8.4               viridisLite_0.4.2          </span></span>
<span><span class="co">#&gt; [155] deldir_2.0-4                BiocParallel_1.38.0        </span></span>
<span><span class="co">#&gt; [157] munsell_0.5.1               lazyeval_0.2.2             </span></span>
<span><span class="co">#&gt; [159] spatstat.geom_3.3-3         CompQuadForm_1.4.3         </span></span>
<span><span class="co">#&gt; [161] RcppHNSW_0.6.0              patchwork_1.3.0            </span></span>
<span><span class="co">#&gt; [163] sparseMatrixStats_1.16.0    future_1.34.0              </span></span>
<span><span class="co">#&gt; [165] statmod_1.5.0               shiny_1.9.1                </span></span>
<span><span class="co">#&gt; [167] highr_0.11                  SummarizedExperiment_1.34.0</span></span>
<span><span class="co">#&gt; [169] ROCR_1.0-11                 broom_1.0.7                </span></span>
<span><span class="co">#&gt; [171] igraph_2.0.3                bslib_0.8.0</span></span></code></pre></div>
</details>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Wei Liu, Yi Yang, Jin Liu.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

      </footer>
</div>






  </body>
</html>
