<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>PRECAST: simulation • PRECAST</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="PRECAST: simulation">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">


    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">PRECAST</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.7</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/PRECAST.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/PRECAST.BreastCancer.html">PRECAST: Human Breast Cancer Data Analysis</a>
    </li>
    <li>
      <a href="../articles/PRECAST.DLPFC.html">PRECAST: DLPFC Single Sample Analysis</a>
    </li>
    <li>
      <a href="../articles/PRECAST.DLPFC4.html">PRECAST: Four DLPFC Sample Analysis</a>
    </li>
    <li>
      <a href="../articles/PRECAST.Simu.html">PRECAST: simulation</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/feiyoung/PRECAST/" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>PRECAST: simulation</h1>
                        <h4 data-toc-skip class="author">Wei Liu</h4>
            
            <h4 data-toc-skip class="date">2025-09-29</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/feiyoung/PRECAST/blob/HEAD/vignettes/PRECAST.Simu.Rmd" class="external-link"><code>vignettes/PRECAST.Simu.Rmd</code></a></small>
      <div class="hidden name"><code>PRECAST.Simu.Rmd</code></div>

    </div>

    
    
<p>This vignette introduces the PRECAST workflow for the analysis of
integrating multiple spatial transcriptomics dataset. The workflow
consists of three steps</p>
<ul>
<li>Independent preprocessing and model setting</li>
<li>Probabilistic embedding, clustering and alignment using PRECAST
model</li>
<li>Downstream analysis (i.e. visualization of clusters and embeddings,
combined differential expression analysis)</li>
</ul>
<p>We demonstrate the use of PRECAST to three simulated Visium data that
are <a href="https://github.com/feiyoung/PRECAST/tree/main/vignettes_data" class="external-link">here</a>,
which can be downloaded to the current working path by the following
command:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">githubURL</span> <span class="op">&lt;-</span> <span class="st">"https://github.com/feiyoung/PRECAST/blob/main/vignettes_data/data_simu.rda?raw=true"</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/download.file.html" class="external-link">download.file</a></span><span class="op">(</span><span class="va">githubURL</span>,<span class="st">"data_simu.rda"</span>,mode<span class="op">=</span><span class="st">'wb'</span><span class="op">)</span></span></code></pre></div>
<p>Then load to R</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/load.html" class="external-link">load</a></span><span class="op">(</span><span class="st">"data_simu.rda"</span><span class="op">)</span></span></code></pre></div>
<p>The package can be loaded with the command:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/feiyoung/PRECAST" class="external-link">PRECAST</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; Loading required package: parallel</span></span>
<span><span class="co">#&gt; Loading required package: gtools</span></span>
<span><span class="co">#&gt; PRECAST :  An efficient data integration method is provided for multiple spatial transcriptomics data with non-cluster-relevant effects such as the complex batch effects. It unifies spatial factor analysis simultaneously with spatial clustering and embedding alignment, requiring only partially shared cell/domain clusters across datasets. More details can be referred to Wei Liu, et al. (2023) &lt;doi:10.1038/s41467-023-35947-w&gt;.   Check out our Package website (https://feiyoung.github.io/PRECAST/index.html) for a more complete description of the methods and analyses</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; Loading required package: SeuratObject</span></span>
<span><span class="co">#&gt; Loading required package: sp</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'SeuratObject'</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:base':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     intersect, t</span></span></code></pre></div>
<div class="section level2">
<h2 id="load-the-simulated-data">Load the simulated data<a class="anchor" aria-label="anchor" href="#load-the-simulated-data"></a>
</h2>
<p>First, we view the the three simulated spatial transcriptomics data
with Visium platform.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">data_simu</span> <span class="co">## a list including three Seurat object with default assay: RNA</span></span>
<span><span class="co">#&gt; [[1]]</span></span>
<span><span class="co">#&gt; An object of class Seurat </span></span>
<span><span class="co">#&gt; 2000 features across 4226 samples within 1 assay </span></span>
<span><span class="co">#&gt; Active assay: RNA (2000 features, 0 variable features)</span></span>
<span><span class="co">#&gt;  2 layers present: counts, data</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; [[2]]</span></span>
<span><span class="co">#&gt; An object of class Seurat </span></span>
<span><span class="co">#&gt; 2000 features across 3661 samples within 1 assay </span></span>
<span><span class="co">#&gt; Active assay: RNA (2000 features, 0 variable features)</span></span>
<span><span class="co">#&gt;  2 layers present: counts, data</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; [[3]]</span></span>
<span><span class="co">#&gt; An object of class Seurat </span></span>
<span><span class="co">#&gt; 2000 features across 3639 samples within 1 assay </span></span>
<span><span class="co">#&gt; Active assay: RNA (2000 features, 0 variable features)</span></span>
<span><span class="co">#&gt;  2 layers present: counts, data</span></span></code></pre></div>
<p>Check the content in <code>data_simu</code>.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">data_simu</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/row.names.html" class="external-link">row.names</a></span><span class="op">(</span><span class="va">data_simu</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="create-a-precastobject-object">Create a PRECASTObject object<a class="anchor" aria-label="anchor" href="#create-a-precastobject-object"></a>
</h2>
<p>We show how to create a PRECASTObject object step by step. First, we
create a Seurat list object using the count matrix and meta data of each
data batch. Although <code>data_simu</code> is a prepared Seurat list
object, we re-create a same objcet seuList to show the details.</p>
<ul>
<li>Note: the spatial coordinates must be contained in the meta data and
named as <code>row</code> and <code>col</code>, which benefits the
identification of spaital coordinates by PRECAST.</li>
</ul>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Get the gene-by-spot read count matrices</span></span>
<span><span class="va">countList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">data_simu</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="va">assay</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.github.io/seurat-object/reference/DefaultAssay.html" class="external-link">DefaultAssay</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://satijalab.github.io/seurat-object/reference/AssayData.html" class="external-link">GetAssayData</a></span><span class="op">(</span><span class="va">x</span>, assay <span class="op">=</span> <span class="va">assay</span>, slot<span class="op">=</span><span class="st">'counts'</span><span class="op">)</span></span>
<span>  </span>
<span><span class="op">}</span> <span class="op">)</span></span>
<span><span class="co">#&gt; Warning: The `slot` argument of `GetAssayData()` is deprecated as of SeuratObject 5.0.0.</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Please use the `layer` argument instead.</span></span>
<span><span class="co">#&gt; <span style="color: #555555;">This warning is displayed once every 8 hours.</span></span></span>
<span><span class="co">#&gt; <span style="color: #555555;">Call `lifecycle::last_lifecycle_warnings()` to see where this warning was</span></span></span>
<span><span class="co">#&gt; <span style="color: #555555;">generated.</span></span></span>
<span></span>
<span><span class="co">## Check the spatial coordinates: Yes, they are named as "row" and "col"!</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">data_simu</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">@</span><span class="va">meta.data</span><span class="op">)</span></span>
<span><span class="co">#&gt;          orig.ident nCount_RNA nFeature_RNA row col sample true_cluster</span></span>
<span><span class="co">#&gt; S1_spot1         S1     179800         1117   0  16     S1            1</span></span>
<span><span class="co">#&gt; S1_spot2         S1     371165         1125  50 102     S1            3</span></span>
<span><span class="co">#&gt; S1_spot3         S1     753086         1108   3  43     S1            1</span></span>
<span><span class="co">#&gt; S1_spot4         S1     133468         1173  59  19     S1            7</span></span>
<span><span class="co">#&gt; S1_spot5         S1     127748         1113  43   9     S1            6</span></span>
<span><span class="co">#&gt; S1_spot6         S1     107114         1131  47  13     S1            6</span></span>
<span></span>
<span><span class="co">## Get the meta data of each spot for each data batch</span></span>
<span><span class="va">metadataList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">data_simu</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="va">x</span><span class="op">@</span><span class="va">meta.data</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="co">## ensure the row.names of metadata in metaList are the same as that of colnames count matrix in countList</span></span>
<span><span class="va">M</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">countList</span><span class="op">)</span></span>
<span><span class="kw">for</span><span class="op">(</span><span class="va">r</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="va">M</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/row.names.html" class="external-link">row.names</a></span><span class="op">(</span><span class="va">metadataList</span><span class="op">[[</span><span class="va">r</span><span class="op">]</span><span class="op">]</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">countList</span><span class="op">[[</span><span class="va">r</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span></span>
<span><span class="co">## Create the Seurat list  object</span></span>
<span></span>
<span><span class="va">seuList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="kw">for</span><span class="op">(</span><span class="va">r</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="va">M</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="va">seuList</span><span class="op">[[</span><span class="va">r</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.github.io/seurat-object/reference/CreateSeuratObject.html" class="external-link">CreateSeuratObject</a></span><span class="op">(</span>counts <span class="op">=</span> <span class="va">countList</span><span class="op">[[</span><span class="va">r</span><span class="op">]</span><span class="op">]</span>, meta.data<span class="op">=</span><span class="va">metadataList</span><span class="op">[[</span><span class="va">r</span><span class="op">]</span><span class="op">]</span>, project <span class="op">=</span> <span class="st">"PRECASTsimu"</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<div class="section level3">
<h3 id="prepare-the-precastobject-with-preprocessing-step-">Prepare the PRECASTObject with preprocessing step.<a class="anchor" aria-label="anchor" href="#prepare-the-precastobject-with-preprocessing-step-"></a>
</h3>
<p>Next, we use <code><a href="../reference/CreateiDRSCObject.html">CreatePRECASTObject()</a></code> to create a
PRECASTObject based on the Seurat list object <code>seuList</code>. This
function will do three things:</p>
<ul>
<li><ol style="list-style-type: decimal">
<li>Filter low-quality spots and genes, controlled by the arguments
<code>premin.features</code> and <code>premin.spots</code>,
respectively; the spots are retained in raw data (seuList) with at least
premin.features number of nonzero-count features (genes), and the genes
are retained in raw data (seuList) with at least
<code>premin.spots</code> number of spots. To ease presentation, we
denote the filtered Seurat list object as data_filter1.</li>
</ol></li>
<li><ol start="2" style="list-style-type: decimal">
<li>Select the top 2,000 variable genes (by setting
<code>gene.number=2000</code>) for each data batch using
<code>FindSVGs()</code> function in <code>DR.SC</code> package for
spatially variable genes or <code><a href="https://satijalab.org/seurat/reference/FindVariableFeatures.html" class="external-link">FindVariableFeatures()</a></code> function
in <code>Seurat</code> package for highly variable genes. Next, we
prioritized genes based on the number of times they were selected as
variable genes in all samples and chose the top 2,000 genes. Then denote
the Seurat list object as data_filter2, where only 2,000 genes are
retained.</li>
</ol></li>
<li><ol start="3" style="list-style-type: decimal">
<li>Conduct strict quality control for data_filter2 by filtering spots
and genes, controlled by the arguments <code>postmin.features</code> and
<code>postmin.spots</code>, respectively; the spots are retained with at
least <code>post.features</code> nonzero counts across genes; the
features (genes) are retained with at least <code>postmin.spots</code>
number of nonzero-count spots. Usually, no genes are filltered because
these genes are variable genes.</li>
</ol></li>
</ul>
<p>If the argument <code>customGenelist</code> is not <code>NULL</code>,
then this function only does (3) based on <code>customGenelist</code>
gene list.</p>
<p>In this simulated dataset, we don’t require to select genes, thus, we
set <code>customGenelist=row.names(seuList[[1]])</code>, representing
the user-defined gene list. User can retain the raw seurat list object
by setting <code>rawData.preserve = TRUE</code>.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co">## Create PRECASTObject</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">2022</span><span class="op">)</span></span>
<span><span class="va">PRECASTObj</span> <span class="op">&lt;-</span>  <span class="fu"><a href="../reference/CreateiDRSCObject.html">CreatePRECASTObject</a></span><span class="op">(</span><span class="va">seuList</span>, customGenelist<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/row.names.html" class="external-link">row.names</a></span><span class="op">(</span><span class="va">seuList</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; Filter spots and features from Raw count data...</span></span>
<span><span class="co">#&gt;  </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; 2025-09-29 12:48:58.026016 : ***** Filtering step for raw count data finished!, 0.04 mins elapsed.</span></span>
<span><span class="co">#&gt; Select the variable genes for each data batch...</span></span>
<span><span class="co">#&gt; CreatePRECASTObject: remove genes:gene252  gene644  gene1235  gene1488  with low count reads in seuList.</span></span>
<span><span class="co">#&gt; 2025-09-29 12:48:58.031165 : ***** Gene selection finished!, 0 mins elapsed.</span></span>
<span><span class="co">#&gt; Filter spots and features from SVGs(HVGs) count data...</span></span>
<span><span class="co">#&gt; Normalizing layer: counts</span></span>
<span><span class="co">#&gt; Normalizing layer: counts</span></span>
<span><span class="co">#&gt; Normalizing layer: counts</span></span>
<span><span class="co">#&gt; 2025-09-29 12:49:04.180414 : ***** Filtering step for count data with variable genes finished!, 0.092 mins elapsed.</span></span>
<span></span>
<span><span class="co">## User can retain the raw seuList by the following commond.</span></span>
<span><span class="co">##  PRECASTObj &lt;-  CreatePRECASTObject(seuList, customGenelist=row.names(seuList[[1]]), rawData.preserve = TRUE)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="fit-precast-using-simulated-data">Fit PRECAST using simulated data<a class="anchor" aria-label="anchor" href="#fit-precast-using-simulated-data"></a>
</h2>
<div class="section level3">
<h3 id="add-the-model-setting">Add the model setting<a class="anchor" aria-label="anchor" href="#add-the-model-setting"></a>
</h3>
<p>Add adjacency matrix list and parameter setting of PRECAST. More
model setting parameters can be found in <code><a href="../reference/model_set.html">model_set()</a></code>.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## check the number of genes/features after filtering step</span></span>
<span><span class="va">PRECASTObj</span><span class="op">@</span><span class="va">seulist</span></span>
<span><span class="co">#&gt; [[1]]</span></span>
<span><span class="co">#&gt; An object of class Seurat </span></span>
<span><span class="co">#&gt; 1996 features across 4226 samples within 1 assay </span></span>
<span><span class="co">#&gt; Active assay: RNA (1996 features, 0 variable features)</span></span>
<span><span class="co">#&gt;  2 layers present: counts, data</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; [[2]]</span></span>
<span><span class="co">#&gt; An object of class Seurat </span></span>
<span><span class="co">#&gt; 1996 features across 3661 samples within 1 assay </span></span>
<span><span class="co">#&gt; Active assay: RNA (1996 features, 0 variable features)</span></span>
<span><span class="co">#&gt;  2 layers present: counts, data</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; [[3]]</span></span>
<span><span class="co">#&gt; An object of class Seurat </span></span>
<span><span class="co">#&gt; 1996 features across 3639 samples within 1 assay </span></span>
<span><span class="co">#&gt; Active assay: RNA (1996 features, 0 variable features)</span></span>
<span><span class="co">#&gt;  2 layers present: counts, data</span></span>
<span></span>
<span><span class="co">## seuList is null since the default value `rawData.preserve` is FALSE.</span></span>
<span><span class="va">PRECASTObj</span><span class="op">@</span><span class="va">seuList</span></span>
<span><span class="co">#&gt; NULL</span></span>
<span></span>
<span><span class="co">## Add adjacency matrix list for a PRECASTObj object to prepare for PRECAST model fitting.</span></span>
<span><span class="va">PRECASTObj</span> <span class="op">&lt;-</span>  <span class="fu"><a href="../reference/AddAdjList.html">AddAdjList</a></span><span class="op">(</span><span class="va">PRECASTObj</span>, platform <span class="op">=</span> <span class="st">"Visium"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Neighbors were identified for 4226 out of 4226 spots.</span></span>
<span><span class="co">#&gt; Neighbors were identified for 3658 out of 3661 spots.</span></span>
<span><span class="co">#&gt; Neighbors were identified for 3638 out of 3639 spots.</span></span>
<span></span>
<span><span class="co">## Add a model setting in advance for a PRECASTObj object: verbose =TRUE helps outputing the information in the algorithm; coreNum set the how many cores are used in PRECAST. If you run PRECAST for multiple number of clusters, you can set multiple cores; otherwise, set it to 1. </span></span>
<span><span class="va">PRECASTObj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/AddParSetting.html">AddParSetting</a></span><span class="op">(</span><span class="va">PRECASTObj</span>, Sigma_equal<span class="op">=</span><span class="cn">FALSE</span>, maxIter<span class="op">=</span><span class="fl">30</span>, verbose<span class="op">=</span><span class="cn">TRUE</span>, int.model<span class="op">=</span><span class="st">'kmeans'</span>,</span>
<span>                             coreNum <span class="op">=</span><span class="fl">1</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="fit-precast">Fit PRECAST<a class="anchor" aria-label="anchor" href="#fit-precast"></a>
</h3>
<p>For function <code>PRECAST</code>, users can specify the number of
clusters
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>K</mi><annotation encoding="application/x-tex">K</annotation></semantics></math>
or set <code>K</code> to be an integer vector by using modified
BIC(MBIC) to determine
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>K</mi><annotation encoding="application/x-tex">K</annotation></semantics></math>.
For convenience, we give a single K here.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">### Given K</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">2022</span><span class="op">)</span></span>
<span><span class="va">PRECASTObj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/PRECAST.html">PRECAST</a></span><span class="op">(</span><span class="va">PRECASTObj</span>, K<span class="op">=</span><span class="fl">7</span><span class="op">)</span></span>
<span><span class="co">#&gt; -----Intergrative data info.: 3 samples, 1996 genes X 11526 spots------</span></span>
<span><span class="co">#&gt; -----Numbers of spots are: 4226, 3661, 3639-----</span></span>
<span><span class="co">#&gt; Starting computing initial values using kmeans ...</span></span>
<span><span class="co">#&gt; Loading required package: irlba</span></span>
<span><span class="co">#&gt; Loading required package: Matrix</span></span>
<span><span class="co">#&gt; 2025-09-29 12:49:31.365267 : ***** Initialization finished!, 0.371 mins elapsed.</span></span>
<span><span class="co">#&gt; ----Fitting PRECAST model...----------------</span></span>
<span><span class="co">#&gt; iter = 2, loglik= 9545476.000000, dloglik=1.004445 </span></span>
<span><span class="co">#&gt; iter = 3, loglik= 10162647.000000, dloglik=0.064656 </span></span>
<span><span class="co">#&gt; iter = 4, loglik= 10206924.000000, dloglik=0.004357 </span></span>
<span><span class="co">#&gt; iter = 5, loglik= 10226364.000000, dloglik=0.001905 </span></span>
<span><span class="co">#&gt; iter = 6, loglik= 10237451.000000, dloglik=0.001084 </span></span>
<span><span class="co">#&gt; iter = 7, loglik= 10244640.000000, dloglik=0.000702 </span></span>
<span><span class="co">#&gt; iter = 8, loglik= 10249816.000000, dloglik=0.000505 </span></span>
<span><span class="co">#&gt; iter = 9, loglik= 10253748.000000, dloglik=0.000384 </span></span>
<span><span class="co">#&gt; iter = 10, loglik= 10257030.000000, dloglik=0.000320 </span></span>
<span><span class="co">#&gt; iter = 11, loglik= 10259924.000000, dloglik=0.000282 </span></span>
<span><span class="co">#&gt; iter = 12, loglik= 10262654.000000, dloglik=0.000266 </span></span>
<span><span class="co">#&gt; iter = 13, loglik= 10265250.000000, dloglik=0.000253 </span></span>
<span><span class="co">#&gt; iter = 14, loglik= 10267834.000000, dloglik=0.000252 </span></span>
<span><span class="co">#&gt; iter = 15, loglik= 10270217.000000, dloglik=0.000232 </span></span>
<span><span class="co">#&gt; iter = 16, loglik= 10272239.000000, dloglik=0.000197 </span></span>
<span><span class="co">#&gt; iter = 17, loglik= 10274072.000000, dloglik=0.000178 </span></span>
<span><span class="co">#&gt; iter = 18, loglik= 10275755.000000, dloglik=0.000164 </span></span>
<span><span class="co">#&gt; iter = 19, loglik= 10277281.000000, dloglik=0.000149 </span></span>
<span><span class="co">#&gt; iter = 20, loglik= 10278724.000000, dloglik=0.000140 </span></span>
<span><span class="co">#&gt; iter = 21, loglik= 10280154.000000, dloglik=0.000139 </span></span>
<span><span class="co">#&gt; iter = 22, loglik= 10281570.000000, dloglik=0.000138 </span></span>
<span><span class="co">#&gt; iter = 23, loglik= 10282945.000000, dloglik=0.000134 </span></span>
<span><span class="co">#&gt; iter = 24, loglik= 10284360.000000, dloglik=0.000138 </span></span>
<span><span class="co">#&gt; iter = 25, loglik= 10285741.000000, dloglik=0.000134 </span></span>
<span><span class="co">#&gt; iter = 26, loglik= 10287149.000000, dloglik=0.000137 </span></span>
<span><span class="co">#&gt; iter = 27, loglik= 10288591.000000, dloglik=0.000140 </span></span>
<span><span class="co">#&gt; iter = 28, loglik= 10289995.000000, dloglik=0.000136 </span></span>
<span><span class="co">#&gt; iter = 29, loglik= 10291425.000000, dloglik=0.000139 </span></span>
<span><span class="co">#&gt; iter = 30, loglik= 10292757.000000, dloglik=0.000129</span></span>
<span><span class="co">#&gt; 2025-09-29 12:59:13.996394 : ***** PRECAST model fitting finished!, 9.71 mins elapsed.</span></span></code></pre></div>
<details><summary><strong>Other options</strong>
</summary><p>Run for multiple K. Here, we set <code>K=6:9</code>.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Reset  parameters by increasing cores.</span></span>
<span><span class="va">PRECASTObj2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/AddParSetting.html">AddParSetting</a></span><span class="op">(</span><span class="va">PRECASTObj</span>, Sigma_equal<span class="op">=</span><span class="cn">FALSE</span>, maxIter<span class="op">=</span><span class="fl">30</span>, verbose<span class="op">=</span><span class="cn">TRUE</span>,</span>
<span>                             coreNum <span class="op">=</span><span class="fl">2</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">2023</span><span class="op">)</span></span>
<span><span class="va">PRECASTObj2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/PRECAST.html">PRECAST</a></span><span class="op">(</span><span class="va">PRECASTObj2</span>, K<span class="op">=</span><span class="fl">6</span><span class="op">:</span><span class="fl">7</span><span class="op">)</span></span>
<span></span>
<span><span class="va">resList2</span> <span class="op">&lt;-</span> <span class="va">PRECASTObj2</span><span class="op">@</span><span class="va">resList</span></span>
<span><span class="va">PRECASTObj2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/selectModel.html">SelectModel</a></span><span class="op">(</span><span class="va">PRECASTObj2</span><span class="op">)</span></span></code></pre></div>
<ul>
<li>Note: For parallel compuation based on Rcpp on Linux, users require
to use the following system command to set the C_stack unlimited in case
of R Error: C stack usage is too close to the limit.</li>
</ul>
<p><code>ulimit -s unlimited</code></p>
<p>Besides, user can also use different initialization method by setting
<code>int.model</code>, for example, set <code>int.model=NULL</code>;
see the functions <code><a href="../reference/AddParSetting.html">AddParSetting()</a></code> and
<code><a href="../reference/model_set.html">model_set()</a></code> for more details.</p>
</details><p>Select a best model and re-organize the results by useing
<code><a href="../reference/selectModel.html">SelectModel()</a></code>. Even though <code>K</code> is not a vector,
it is also necessary to run <code><a href="../reference/selectModel.html">SelectModel()</a></code> to re-organize
the results in <code>PRECASTObj</code>. The selected best K is 7 by
using command <code>str(PRECASTObj@resList)</code>.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## check the fitted results: there are four list for the fitted results of each K (6:9).</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">PRECASTObj</span><span class="op">@</span><span class="va">resList</span><span class="op">)</span></span>
<span><span class="co">#&gt; List of 1</span></span>
<span><span class="co">#&gt;  $ :List of 12</span></span>
<span><span class="co">#&gt;   ..$ cluster   :List of 3</span></span>
<span><span class="co">#&gt;   .. ..$ : num [1:4226, 1] 4 1 4 5 4 4 5 3 4 1 ...</span></span>
<span><span class="co">#&gt;   .. ..$ : num [1:3661, 1] 3 4 1 3 4 1 1 3 4 5 ...</span></span>
<span><span class="co">#&gt;   .. ..$ : num [1:3639, 1] 4 4 5 1 4 4 5 1 4 1 ...</span></span>
<span><span class="co">#&gt;   .. ..- attr(*, "dim")= int [1:2] 3 1</span></span>
<span><span class="co">#&gt;   ..$ hZ        :List of 3</span></span>
<span><span class="co">#&gt;   .. ..$ : num [1:4226, 1:15] -0.845 -0.464 -0.845 -0.722 -0.822 ...</span></span>
<span><span class="co">#&gt;   .. ..$ : num [1:3661, 1:15] -0.975 -0.981 -0.451 -1.052 -0.846 ...</span></span>
<span><span class="co">#&gt;   .. ..$ : num [1:3639, 1:15] -0.866 -0.928 -0.707 -0.424 -0.955 ...</span></span>
<span><span class="co">#&gt;   .. ..- attr(*, "dim")= int [1:2] 3 1</span></span>
<span><span class="co">#&gt;   ..$ hV        :List of 3</span></span>
<span><span class="co">#&gt;   .. ..$ : num [1:4226, 1:15] 3.25 10.48 18.34 -3.52 -4.02 ...</span></span>
<span><span class="co">#&gt;   .. ..$ : num [1:3661, 1:15] 13.63 -7.99 -5.71 -2.61 2.19 ...</span></span>
<span><span class="co">#&gt;   .. ..$ : num [1:3639, 1:15] 5.262 -0.471 -9.159 0.259 -0.179 ...</span></span>
<span><span class="co">#&gt;   .. ..- attr(*, "dim")= int [1:2] 3 1</span></span>
<span><span class="co">#&gt;   ..$ Rf        :List of 3</span></span>
<span><span class="co">#&gt;   .. ..$ : num [1:4226, 1:7] 3.71e-05 1.00 6.29e-06 6.67e-25 1.98e-05 ...</span></span>
<span><span class="co">#&gt;   .. ..$ : num [1:3661, 1:7] 6.60e-04 3.80e-04 1.00 1.13e-06 6.08e-07 ...</span></span>
<span><span class="co">#&gt;   .. ..$ : num [1:3639, 1:7] 7.36e-03 1.81e-04 2.67e-11 1.00 5.01e-03 ...</span></span>
<span><span class="co">#&gt;   .. ..- attr(*, "dim")= int [1:2] 3 1</span></span>
<span><span class="co">#&gt;   ..$ beta      : num [1:3, 1] 2.8 4 3.6</span></span>
<span><span class="co">#&gt;   ..$ Mu        : num [1:7, 1:15] -0.479 -1.67 -1.048 -0.861 -0.691 ...</span></span>
<span><span class="co">#&gt;   ..$ Sigma     : num [1:15, 1:15, 1:7] 0.0234 0 0 0 0 ...</span></span>
<span><span class="co">#&gt;   ..$ Psi       : num [1:15, 1:15, 1:3] 82.0084 -28.1893 2.1016 0.0417 0.7358 ...</span></span>
<span><span class="co">#&gt;   ..$ W         : num [1:1996, 1:15] -0.01287 -0.00644 -0.01465 -0.02239 -0.008 ...</span></span>
<span><span class="co">#&gt;   ..$ Lam       : num [1:3, 1:1996] 0.2362 1.6618 0.2006 0.015 0.0237 ...</span></span>
<span><span class="co">#&gt;   ..$ loglik    : num 10292757</span></span>
<span><span class="co">#&gt;   ..$ loglik_seq: num [1:30, 1] -2.15e+09 9.55e+06 1.02e+07 1.02e+07 1.02e+07 ...</span></span>
<span><span class="co">#&gt;  - attr(*, "para_settings")=List of 8</span></span>
<span><span class="co">#&gt;   ..$ K             : num 7</span></span>
<span><span class="co">#&gt;   ..$ n             : int 11526</span></span>
<span><span class="co">#&gt;   ..$ p             : int 1996</span></span>
<span><span class="co">#&gt;   ..$ q             : num 15</span></span>
<span><span class="co">#&gt;   ..$ r_max         : int 3</span></span>
<span><span class="co">#&gt;   ..$ Sigma_equal   : logi FALSE</span></span>
<span><span class="co">#&gt;   ..$ Sigma_diag    : logi TRUE</span></span>
<span><span class="co">#&gt;   ..$ mix_prop_heter: logi TRUE</span></span>
<span><span class="co">#&gt;  - attr(*, "class")= chr "SeqK_PRECAST_Object"</span></span>
<span><span class="co">## backup the fitted results in resList</span></span>
<span><span class="va">resList</span> <span class="op">&lt;-</span> <span class="va">PRECASTObj</span><span class="op">@</span><span class="va">resList</span></span>
<span><span class="co"># PRECASTObj@resList &lt;- resList</span></span>
<span><span class="va">PRECASTObj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/selectModel.html">SelectModel</a></span><span class="op">(</span><span class="va">PRECASTObj</span><span class="op">)</span></span>
<span><span class="co">## check the best and re-organized results</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">PRECASTObj</span><span class="op">@</span><span class="va">resList</span><span class="op">)</span> <span class="co">## The selected best K is 7</span></span>
<span><span class="co">#&gt; List of 7</span></span>
<span><span class="co">#&gt;  $ bestK  : num 7</span></span>
<span><span class="co">#&gt;  $ cluster:List of 3</span></span>
<span><span class="co">#&gt;   ..$ : num [1:4226, 1] 4 1 4 5 4 4 5 3 4 1 ...</span></span>
<span><span class="co">#&gt;   ..$ : num [1:3661, 1] 3 4 1 3 4 1 1 3 4 5 ...</span></span>
<span><span class="co">#&gt;   ..$ : num [1:3639, 1] 4 4 5 1 4 4 5 1 4 1 ...</span></span>
<span><span class="co">#&gt;   ..- attr(*, "dim")= int [1:2] 3 1</span></span>
<span><span class="co">#&gt;  $ hZ     :List of 3</span></span>
<span><span class="co">#&gt;   ..$ : num [1:4226, 1:15] -0.845 -0.464 -0.845 -0.722 -0.822 ...</span></span>
<span><span class="co">#&gt;   ..$ : num [1:3661, 1:15] -0.975 -0.981 -0.451 -1.052 -0.846 ...</span></span>
<span><span class="co">#&gt;   ..$ : num [1:3639, 1:15] -0.866 -0.928 -0.707 -0.424 -0.955 ...</span></span>
<span><span class="co">#&gt;   ..- attr(*, "dim")= int [1:2] 3 1</span></span>
<span><span class="co">#&gt;  $ Rf     :List of 3</span></span>
<span><span class="co">#&gt;   ..$ : num [1:4226, 1:7] 3.71e-05 1.00 6.29e-06 6.67e-25 1.98e-05 ...</span></span>
<span><span class="co">#&gt;   ..$ : num [1:3661, 1:7] 6.60e-04 3.80e-04 1.00 1.13e-06 6.08e-07 ...</span></span>
<span><span class="co">#&gt;   ..$ : num [1:3639, 1:7] 7.36e-03 1.81e-04 2.67e-11 1.00 5.01e-03 ...</span></span>
<span><span class="co">#&gt;   ..- attr(*, "dim")= int [1:2] 3 1</span></span>
<span><span class="co">#&gt;  $ hV     :List of 3</span></span>
<span><span class="co">#&gt;   ..$ : num [1:4226, 1:15] 3.25 10.48 18.34 -3.52 -4.02 ...</span></span>
<span><span class="co">#&gt;   ..$ : num [1:3661, 1:15] 13.63 -7.99 -5.71 -2.61 2.19 ...</span></span>
<span><span class="co">#&gt;   ..$ : num [1:3639, 1:15] 5.262 -0.471 -9.159 0.259 -0.179 ...</span></span>
<span><span class="co">#&gt;   ..- attr(*, "dim")= int [1:2] 3 1</span></span>
<span><span class="co">#&gt;  $ hW     : num [1:1996, 1:15] -0.01287 -0.00644 -0.01465 -0.02239 -0.008 ...</span></span>
<span><span class="co">#&gt;  $ icMat  : num [1, 1:2] 7 -19816556</span></span>
<span><span class="co">#&gt;   ..- attr(*, "dimnames")=List of 2</span></span>
<span><span class="co">#&gt;   .. ..$ : NULL</span></span>
<span><span class="co">#&gt;   .. ..$ : chr [1:2] "K" "IC"</span></span></code></pre></div>
<p>Use ARI to check the performance of clustering:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">true_cluster</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">PRECASTObj</span><span class="op">@</span><span class="va">seulist</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="va">x</span><span class="op">$</span><span class="va">true_cluster</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">true_cluster</span><span class="op">)</span></span>
<span><span class="co">#&gt; List of 3</span></span>
<span><span class="co">#&gt;  $ : Named num [1:4226] 1 3 1 7 6 6 7 5 3 3 ...</span></span>
<span><span class="co">#&gt;   ..- attr(*, "names")= chr [1:4226] "S1_spot1" "S1_spot2" "S1_spot3" "S1_spot4" ...</span></span>
<span><span class="co">#&gt;  $ : Named num [1:3661] 4 6 3 4 6 3 3 4 6 7 ...</span></span>
<span><span class="co">#&gt;   ..- attr(*, "names")= chr [1:3661] "S1_spot1" "S1_spot2" "S1_spot3" "S1_spot4" ...</span></span>
<span><span class="co">#&gt;  $ : Named num [1:3639] 3 1 7 3 5 6 7 3 2 6 ...</span></span>
<span><span class="co">#&gt;   ..- attr(*, "names")= chr [1:3639] "S1_spot1" "S1_spot2" "S1_spot3" "S1_spot4" ...</span></span>
<span><span class="fu">mclust</span><span class="fu">::</span><span class="fu"><a href="https://mclust-org.github.io/mclust/reference/adjustedRandIndex.html" class="external-link">adjustedRandIndex</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="va">PRECASTObj</span><span class="op">@</span><span class="va">resList</span><span class="op">$</span><span class="va">cluster</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="va">true_cluster</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 0.5787175</span></span></code></pre></div>
<p>We provide two methods to correct the batch effects in gene
expression level. Method (1) is using only PRECAST results to obtain the
batch corrected gene expressions if the species of data is unknown or
the number of overlapped housekeeping genes between the variable genes
in <code>PRECASTObj@seulist</code> and the genes in database is less
than five. Method (2) is using bouth housekeeping gene and PRECAST
results to obtain the batch corrected gene expressions.</p>
<ul>
<li>Note: to obtain batch corrected gene expressions based on
housekeeping genes as the negative control, users must specify the
species of data source and use gene symbol names in
<code>PRECASTObj@seulist</code>.</li>
</ul>
<p>Integrate the two samples by the function
<code>IntegrateSpaData</code>. Because this is a simulated data, we use
Method (1) by setting <code>species='unknown'</code>.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">seuInt</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/IntegrateSpaData.html">IntegrateSpaData</a></span><span class="op">(</span><span class="va">PRECASTObj</span>, species<span class="op">=</span><span class="st">'unknown'</span><span class="op">)</span></span>
<span><span class="co">#&gt; Using only PRECAST results to obtain the batch corrected gene expressions since species is unknown or the genelist in PRECASTObj has less than 5 overlapp with the housekeeping genes of given species.</span></span>
<span><span class="co">#&gt; Start integration...</span></span>
<span><span class="co">#&gt; 2025-09-29 12:59:20.776706 : ***** Data integration finished!, 0.096 mins elapsed.</span></span>
<span><span class="co">#&gt; Put the data into a new Seurat object...</span></span>
<span><span class="co">#&gt; 2025-09-29 12:59:22.661266 : ***** New Seurat object is generated!, 0.031 mins elapsed.</span></span>
<span><span class="va">seuInt</span> </span>
<span><span class="co">#&gt; An object of class Seurat </span></span>
<span><span class="co">#&gt; 1996 features across 11526 samples within 1 assay </span></span>
<span><span class="co">#&gt; Active assay: PRE_CAST (1996 features, 0 variable features)</span></span>
<span><span class="co">#&gt;  2 layers present: counts, data</span></span>
<span><span class="co">#&gt;  2 dimensional reductions calculated: PRECAST, position</span></span>
<span><span class="co">## The low-dimensional embeddings obtained by PRECAST are saved in PRECAST reduction slot.</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="visualization">Visualization<a class="anchor" aria-label="anchor" href="#visualization"></a>
</h2>
<p>First, user can choose a beautiful color schema using
<code><a href="../reference/chooseColors.html">chooseColors()</a></code>.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cols_cluster</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/chooseColors.html">chooseColors</a></span><span class="op">(</span>palettes_name <span class="op">=</span> <span class="st">'Nature 10'</span>, n_colors <span class="op">=</span> <span class="fl">7</span>, plot_colors <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="PRECAST.Simu_files/figure-html/unnamed-chunk-16-1.png" width="700"></p>
<p>Show the spatial scatter plot for clusters</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p12</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/SpaPlot.html">SpaPlot</a></span><span class="op">(</span><span class="va">seuInt</span>, batch<span class="op">=</span><span class="cn">NULL</span>, cols<span class="op">=</span><span class="va">cols_cluster</span>, point_size<span class="op">=</span><span class="fl">2</span>, combine<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">p12</span></span></code></pre></div>
<p><img src="PRECAST.Simu_files/figure-html/unnamed-chunk-17-1.png" width="672"></p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># users can plot each sample by setting combine=FALSE</span></span></code></pre></div>
<p>Users can re-plot the above figures for specific need by returning a
ggplot list object. For example, we only plot the spatial heatmap of
first two data batches.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pList</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/SpaPlot.html">SpaPlot</a></span><span class="op">(</span><span class="va">seuInt</span>, batch<span class="op">=</span><span class="cn">NULL</span>, cols<span class="op">=</span><span class="va">cols_cluster</span>, point_size<span class="op">=</span><span class="fl">2</span>, combine<span class="op">=</span><span class="cn">FALSE</span>, title_name<span class="op">=</span><span class="cn">NULL</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/drawFigs.html">drawFigs</a></span><span class="op">(</span><span class="va">pList</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">]</span>, layout.dim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">2</span><span class="op">)</span>, common.legend <span class="op">=</span> <span class="cn">TRUE</span>, legend.position <span class="op">=</span> <span class="st">'right'</span>, align<span class="op">=</span><span class="st">'hv'</span><span class="op">)</span></span></code></pre></div>
<p><img src="PRECAST.Simu_files/figure-html/unnamed-chunk-18-1.png" width="672"></p>
<p>Show the spatial UMAP/tNSE RGB plot</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seuInt</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/AddUMAP.html">AddUMAP</a></span><span class="op">(</span><span class="va">seuInt</span><span class="op">)</span> </span>
<span><span class="fu"><a href="../reference/SpaPlot.html">SpaPlot</a></span><span class="op">(</span><span class="va">seuInt</span>, batch<span class="op">=</span><span class="cn">NULL</span>,item<span class="op">=</span><span class="st">'RGB_UMAP'</span>,point_size<span class="op">=</span><span class="fl">1</span>, combine<span class="op">=</span><span class="cn">TRUE</span>, text_size<span class="op">=</span><span class="fl">15</span><span class="op">)</span></span></code></pre></div>
<p><img src="PRECAST.Simu_files/figure-html/unnamed-chunk-19-1.png" width="576"></p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co">## Plot tSNE RGB plot</span></span>
<span><span class="co">#seuInt &lt;- AddTSNE(seuInt) </span></span>
<span><span class="co">#SpaPlot(seuInt, batch=NULL,item='RGB_TSNE',point_size=2, combine=T, text_size=15)</span></span></code></pre></div>
<p>Show the tSNE plot based on the extracted features from PRECAST to
check the performance of integration.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seuInt</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/AddTSNE.html">AddTSNE</a></span><span class="op">(</span><span class="va">seuInt</span>, n_comp <span class="op">=</span> <span class="fl">2</span><span class="op">)</span> </span>
<span></span>
<span><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/dimPlot.html">dimPlot</a></span><span class="op">(</span><span class="va">seuInt</span>, item<span class="op">=</span><span class="st">'cluster'</span>, font_family<span class="op">=</span><span class="st">'serif'</span>, cols<span class="op">=</span><span class="va">cols_cluster</span><span class="op">)</span> <span class="co"># Times New Roman</span></span>
<span><span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/dimPlot.html">dimPlot</a></span><span class="op">(</span><span class="va">seuInt</span>, item<span class="op">=</span><span class="st">'batch'</span>, point_size <span class="op">=</span> <span class="fl">1</span>,  font_family<span class="op">=</span><span class="st">'serif'</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/drawFigs.html">drawFigs</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">p1</span>, <span class="va">p2</span><span class="op">)</span>, common.legend<span class="op">=</span><span class="cn">FALSE</span>, align<span class="op">=</span><span class="st">'hv'</span><span class="op">)</span> </span></code></pre></div>
<p><img src="PRECAST.Simu_files/figure-html/unnamed-chunk-20-1.png" width="576"></p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># It is noted that only sample batch 1 has cluster 4, and only sample batch 2 has cluster 7. </span></span></code></pre></div>
<p>Show the UMAP plot based on the extracted features from PRECAST.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/dimPlot.html">dimPlot</a></span><span class="op">(</span><span class="va">seuInt</span>, reduction <span class="op">=</span> <span class="st">'UMAP3'</span>, item<span class="op">=</span><span class="st">'cluster'</span>, cols<span class="op">=</span><span class="va">cols_cluster</span>, font_family<span class="op">=</span><span class="st">'serif'</span><span class="op">)</span></span></code></pre></div>
<p><img src="PRECAST.Simu_files/figure-html/unnamed-chunk-21-1.png" width="576"></p>
<p>Users can also use the visualization functions in Seurat package:</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span></span>
<span><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">seuInt</span><span class="op">[</span>,<span class="fl">1</span><span class="op">:</span> <span class="fl">4226</span><span class="op">]</span>, reduction <span class="op">=</span> <span class="st">'position'</span>, cols<span class="op">=</span><span class="va">cols_cluster</span>, pt.size <span class="op">=</span><span class="fl">1</span><span class="op">)</span> <span class="co"># plot the first data batch: first 4226 spots.</span></span>
<span><span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">seuInt</span>, reduction <span class="op">=</span> <span class="st">'tSNE'</span>,cols<span class="op">=</span><span class="va">cols_cluster</span>, pt.size<span class="op">=</span><span class="fl">1</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/drawFigs.html">drawFigs</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">p1</span>, <span class="va">p2</span><span class="op">)</span>, layout.dim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">2</span><span class="op">)</span>, common.legend <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="PRECAST.Simu_files/figure-html/unnamed-chunk-22-1.png" width="768"></p>
<p>Combined differential expression analysis</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dat_deg</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindAllMarkers.html" class="external-link">FindAllMarkers</a></span><span class="op">(</span><span class="va">seuInt</span><span class="op">)</span></span>
<span><span class="co">#&gt; Calculating cluster 1</span></span>
<span><span class="co">#&gt; Calculating cluster 2</span></span>
<span><span class="co">#&gt; Calculating cluster 3</span></span>
<span><span class="co">#&gt; Calculating cluster 4</span></span>
<span><span class="co">#&gt; Calculating cluster 5</span></span>
<span><span class="co">#&gt; Calculating cluster 6</span></span>
<span><span class="co">#&gt; Calculating cluster 7</span></span>
<span><span class="co">#&gt; Warning in mean.fxn(object[features, cells.1, drop = FALSE]): NaNs produced</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'dplyr'</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:stats':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     filter, lag</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:base':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     intersect, setdiff, setequal, union</span></span>
<span><span class="va">n</span> <span class="op">&lt;-</span> <span class="fl">2</span></span>
<span><span class="va">dat_deg</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/dplyr/man/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">cluster</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/dplyr/man/top_n.html" class="external-link">top_n</a></span><span class="op">(</span>n <span class="op">=</span> <span class="va">n</span>, wt <span class="op">=</span> <span class="va">avg_log2FC</span><span class="op">)</span> <span class="op">-&gt;</span> <span class="va">top10</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">top10</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 6 × 7</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># Groups:   cluster [3]</span></span></span>
<span><span class="co">#&gt;      p_val avg_log2FC pct.1 pct.2 p_val_adj cluster gene    </span></span>
<span><span class="co">#&gt;      <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> 6.12<span style="color: #949494;">e</span><span style="color: #BB0000;">-98</span>       1.63 0.859 0.803  1.22<span style="color: #949494;">e</span><span style="color: #BB0000;">-94</span> 1       gene1953</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> 3.41<span style="color: #949494;">e</span><span style="color: #BB0000;">-24</span>       2.40 0.629 0.702  6.81<span style="color: #949494;">e</span><span style="color: #BB0000;">-21</span> 1       gene670 </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> 1.77<span style="color: #949494;">e</span><span style="color: #BB0000;">- 3</span>       2.68 0.706 0.765  1   <span style="color: #949494;">e</span>+ 0 2       gene1142</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> 4.64<span style="color: #949494;">e</span><span style="color: #BB0000;">- 3</span>       2.26 0.831 0.827  1   <span style="color: #949494;">e</span>+ 0 2       gene1237</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span> 2.24<span style="color: #949494;">e</span><span style="color: #BB0000;">- 6</span>       1.80 0.716 0.657  4.47<span style="color: #949494;">e</span><span style="color: #BB0000;">- 3</span> 3       gene274 </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">6</span> 5.25<span style="color: #949494;">e</span><span style="color: #BB0000;">- 5</span>       2.46 0.736 0.635  1.05<span style="color: #949494;">e</span><span style="color: #BB0000;">- 1</span> 3       gene948</span></span></code></pre></div>
<details><summary><strong>Session Info</strong>
</summary><div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.4.1 (2024-06-14 ucrt)</span></span>
<span><span class="co">#&gt; Platform: x86_64-w64-mingw32/x64</span></span>
<span><span class="co">#&gt; Running under: Windows 11 x64 (build 26100)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt; [1] LC_COLLATE=Chinese (Simplified)_China.utf8 </span></span>
<span><span class="co">#&gt; [2] LC_CTYPE=Chinese (Simplified)_China.utf8   </span></span>
<span><span class="co">#&gt; [3] LC_MONETARY=Chinese (Simplified)_China.utf8</span></span>
<span><span class="co">#&gt; [4] LC_NUMERIC=C                               </span></span>
<span><span class="co">#&gt; [5] LC_TIME=Chinese (Simplified)_China.utf8    </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: Asia/Shanghai</span></span>
<span><span class="co">#&gt; tzcode source: internal</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] parallel  stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">#&gt; [8] base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt; [1] dplyr_1.1.4        irlba_2.3.5.1      Matrix_1.7-0       Seurat_5.1.0      </span></span>
<span><span class="co">#&gt; [5] SeuratObject_5.0.2 sp_2.1-4           PRECAST_1.7        gtools_3.9.5      </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;   [1] RcppAnnoy_0.0.22            splines_4.4.1              </span></span>
<span><span class="co">#&gt;   [3] later_1.3.2                 tibble_3.2.1               </span></span>
<span><span class="co">#&gt;   [5] polyclip_1.10-7             fastDummies_1.7.4          </span></span>
<span><span class="co">#&gt;   [7] lifecycle_1.0.4             rstatix_0.7.2              </span></span>
<span><span class="co">#&gt;   [9] globals_0.16.3              lattice_0.22-6             </span></span>
<span><span class="co">#&gt;  [11] MASS_7.3-60.2               backports_1.5.0            </span></span>
<span><span class="co">#&gt;  [13] magrittr_2.0.3              limma_3.58.1               </span></span>
<span><span class="co">#&gt;  [15] plotly_4.10.4               sass_0.4.9                 </span></span>
<span><span class="co">#&gt;  [17] rmarkdown_2.28              jquerylib_0.1.4            </span></span>
<span><span class="co">#&gt;  [19] yaml_2.3.10                 httpuv_1.6.15              </span></span>
<span><span class="co">#&gt;  [21] sctransform_0.4.1           spam_2.10-0                </span></span>
<span><span class="co">#&gt;  [23] spatstat.sparse_3.1-0       reticulate_1.39.0          </span></span>
<span><span class="co">#&gt;  [25] cowplot_1.1.3               pbapply_1.7-2              </span></span>
<span><span class="co">#&gt;  [27] RColorBrewer_1.1-3          abind_1.4-8                </span></span>
<span><span class="co">#&gt;  [29] zlibbioc_1.50.0             Rtsne_0.17                 </span></span>
<span><span class="co">#&gt;  [31] GenomicRanges_1.56.2        presto_1.0.0               </span></span>
<span><span class="co">#&gt;  [33] purrr_1.0.2                 BiocGenerics_0.50.0        </span></span>
<span><span class="co">#&gt;  [35] GenomeInfoDbData_1.2.12     IRanges_2.38.1             </span></span>
<span><span class="co">#&gt;  [37] S4Vectors_0.42.1            ggrepel_0.9.6              </span></span>
<span><span class="co">#&gt;  [39] listenv_0.9.1               spatstat.utils_3.1-0       </span></span>
<span><span class="co">#&gt;  [41] goftest_1.2-3               RSpectra_0.16-2            </span></span>
<span><span class="co">#&gt;  [43] spatstat.random_3.3-2       fitdistrplus_1.2-1         </span></span>
<span><span class="co">#&gt;  [45] parallelly_1.38.0           pkgdown_2.1.1              </span></span>
<span><span class="co">#&gt;  [47] DelayedMatrixStats_1.26.0   leiden_0.4.3.1             </span></span>
<span><span class="co">#&gt;  [49] codetools_0.2-20            DelayedArray_0.30.1        </span></span>
<span><span class="co">#&gt;  [51] scuttle_1.14.0              tidyselect_1.2.1           </span></span>
<span><span class="co">#&gt;  [53] UCSC.utils_1.0.0            farver_2.1.2               </span></span>
<span><span class="co">#&gt;  [55] viridis_0.6.5               ScaledMatrix_1.12.0        </span></span>
<span><span class="co">#&gt;  [57] matrixStats_1.4.1           stats4_4.4.1               </span></span>
<span><span class="co">#&gt;  [59] spatstat.explore_3.3-2      jsonlite_1.8.9             </span></span>
<span><span class="co">#&gt;  [61] BiocNeighbors_1.22.0        Formula_1.2-5              </span></span>
<span><span class="co">#&gt;  [63] progressr_0.14.0            ggridges_0.5.6             </span></span>
<span><span class="co">#&gt;  [65] survival_3.6-4              scater_1.32.1              </span></span>
<span><span class="co">#&gt;  [67] systemfonts_1.1.0           tools_4.4.1                </span></span>
<span><span class="co">#&gt;  [69] ragg_1.3.3                  ica_1.0-3                  </span></span>
<span><span class="co">#&gt;  [71] Rcpp_1.0.13                 glue_1.7.0                 </span></span>
<span><span class="co">#&gt;  [73] gridExtra_2.3               SparseArray_1.4.8          </span></span>
<span><span class="co">#&gt;  [75] xfun_0.47                   MatrixGenerics_1.16.0      </span></span>
<span><span class="co">#&gt;  [77] ggthemes_5.1.0              GenomeInfoDb_1.40.1        </span></span>
<span><span class="co">#&gt;  [79] withr_3.0.1                 fastmap_1.2.0              </span></span>
<span><span class="co">#&gt;  [81] fansi_1.0.6                 digest_0.6.37              </span></span>
<span><span class="co">#&gt;  [83] rsvd_1.0.5                  R6_2.5.1                   </span></span>
<span><span class="co">#&gt;  [85] mime_0.12                   textshaping_0.4.0          </span></span>
<span><span class="co">#&gt;  [87] colorspace_2.1-1            scattermore_1.2            </span></span>
<span><span class="co">#&gt;  [89] tensor_1.5                  spatstat.data_3.1-2        </span></span>
<span><span class="co">#&gt;  [91] RhpcBLASctl_0.23-42         utf8_1.2.4                 </span></span>
<span><span class="co">#&gt;  [93] tidyr_1.3.1                 generics_0.1.3             </span></span>
<span><span class="co">#&gt;  [95] data.table_1.16.0           httr_1.4.7                 </span></span>
<span><span class="co">#&gt;  [97] htmlwidgets_1.6.4           S4Arrays_1.4.1             </span></span>
<span><span class="co">#&gt;  [99] uwot_0.2.2                  pkgconfig_2.0.3            </span></span>
<span><span class="co">#&gt; [101] gtable_0.3.5                lmtest_0.9-40              </span></span>
<span><span class="co">#&gt; [103] SingleCellExperiment_1.26.0 XVector_0.44.0             </span></span>
<span><span class="co">#&gt; [105] htmltools_0.5.8.1           carData_3.0-5              </span></span>
<span><span class="co">#&gt; [107] dotCall64_1.1-1             scales_1.3.0               </span></span>
<span><span class="co">#&gt; [109] Biobase_2.64.0              png_0.1-8                  </span></span>
<span><span class="co">#&gt; [111] harmony_1.2.1               spatstat.univar_3.0-1      </span></span>
<span><span class="co">#&gt; [113] knitr_1.48                  rstudioapi_0.16.0          </span></span>
<span><span class="co">#&gt; [115] reshape2_1.4.4              nlme_3.1-164               </span></span>
<span><span class="co">#&gt; [117] cachem_1.1.0                zoo_1.8-12                 </span></span>
<span><span class="co">#&gt; [119] stringr_1.5.1               KernSmooth_2.23-24         </span></span>
<span><span class="co">#&gt; [121] vipor_0.4.7                 miniUI_0.1.1.1             </span></span>
<span><span class="co">#&gt; [123] GiRaF_1.0.1                 desc_1.4.3                 </span></span>
<span><span class="co">#&gt; [125] pillar_1.9.0                grid_4.4.1                 </span></span>
<span><span class="co">#&gt; [127] vctrs_0.6.5                 RANN_2.6.2                 </span></span>
<span><span class="co">#&gt; [129] ggpubr_0.6.0                promises_1.3.0             </span></span>
<span><span class="co">#&gt; [131] car_3.1-3                   BiocSingular_1.20.0        </span></span>
<span><span class="co">#&gt; [133] DR.SC_3.4                   beachmat_2.20.0            </span></span>
<span><span class="co">#&gt; [135] xtable_1.8-4                cluster_2.1.6              </span></span>
<span><span class="co">#&gt; [137] beeswarm_0.4.0              evaluate_1.0.0             </span></span>
<span><span class="co">#&gt; [139] cli_3.6.3                   compiler_4.4.1             </span></span>
<span><span class="co">#&gt; [141] rlang_1.1.4                 crayon_1.5.3               </span></span>
<span><span class="co">#&gt; [143] ggsignif_0.6.4              future.apply_1.11.2        </span></span>
<span><span class="co">#&gt; [145] labeling_0.4.3              mclust_6.1.1               </span></span>
<span><span class="co">#&gt; [147] plyr_1.8.9                  fs_1.6.4                   </span></span>
<span><span class="co">#&gt; [149] ggbeeswarm_0.7.2            stringi_1.8.4              </span></span>
<span><span class="co">#&gt; [151] viridisLite_0.4.2           deldir_2.0-4               </span></span>
<span><span class="co">#&gt; [153] BiocParallel_1.38.0         munsell_0.5.1              </span></span>
<span><span class="co">#&gt; [155] lazyeval_0.2.2              spatstat.geom_3.3-3        </span></span>
<span><span class="co">#&gt; [157] CompQuadForm_1.4.3          RcppHNSW_0.6.0             </span></span>
<span><span class="co">#&gt; [159] patchwork_1.3.0             sparseMatrixStats_1.16.0   </span></span>
<span><span class="co">#&gt; [161] future_1.34.0               ggplot2_3.5.2              </span></span>
<span><span class="co">#&gt; [163] statmod_1.5.0               shiny_1.9.1                </span></span>
<span><span class="co">#&gt; [165] highr_0.11                  SummarizedExperiment_1.34.0</span></span>
<span><span class="co">#&gt; [167] ROCR_1.0-11                 broom_1.0.7                </span></span>
<span><span class="co">#&gt; [169] igraph_2.0.3                bslib_0.8.0</span></span></code></pre></div>
</details>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Wei Liu, Yi Yang, Jin Liu.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

      </footer>
</div>






  </body>
</html>
