<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>ICM-EM algorithm implementation — ICM.EM • PRECAST</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="ICM-EM algorithm implementation — ICM.EM"><meta property="og:description" content="ICM-EM algorithm for fitting PRECAST model"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">


    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">PRECAST</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.7</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../articles/PRECAST.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/PRECAST.BreastCancer.html">PRECAST: Human Breast Cancer Data Analysis</a>
    </li>
    <li>
      <a href="../articles/PRECAST.DLPFC.html">PRECAST: DLPFC Single Sample Analysis</a>
    </li>
    <li>
      <a href="../articles/PRECAST.DLPFC4.html">PRECAST: Four DLPFC Sample Analysis</a>
    </li>
    <li>
      <a href="../articles/PRECAST.Simu.html">PRECAST: simulation</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/feiyoung/PRECAST/" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>ICM-EM algorithm implementation</h1>

    <div class="hidden name"><code>idrsc.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>ICM-EM algorithm for fitting PRECAST model</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">ICM.EM</span><span class="op">(</span><span class="va">XList</span>, <span class="va">q</span>, <span class="va">K</span>, AdjList<span class="op">=</span><span class="cn">NULL</span>,  Adjlist_car<span class="op">=</span><span class="cn">NULL</span>, posList <span class="op">=</span> <span class="cn">NULL</span>, platform <span class="op">=</span> <span class="st">"ST"</span>,</span>
<span>         beta_grid<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">0.2</span>,<span class="fl">4</span>, by<span class="op">=</span><span class="fl">0.2</span><span class="op">)</span>, init.nstart<span class="op">=</span><span class="fl">5</span>, </span>
<span>         maxIter_ICM<span class="op">=</span><span class="fl">6</span>,maxIter<span class="op">=</span><span class="fl">20</span>, epsLogLik<span class="op">=</span><span class="fl">1e-5</span>, verbose<span class="op">=</span><span class="cn">TRUE</span>,</span>
<span>         mix_prop_heter<span class="op">=</span><span class="cn">TRUE</span>, Sigma_equal<span class="op">=</span><span class="cn">FALSE</span>, Sigma_diag<span class="op">=</span><span class="cn">TRUE</span>,error_heter<span class="op">=</span><span class="cn">TRUE</span>, Sp2<span class="op">=</span><span class="cn">TRUE</span>,</span>
<span>         wpca_int<span class="op">=</span><span class="cn">FALSE</span>,int.model<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'kmeans'</span>, <span class="st">'mclust'</span><span class="op">)</span>, seed<span class="op">=</span><span class="fl">1</span>,coreNum <span class="op">=</span> <span class="fl">1</span>, coreNum_int<span class="op">=</span><span class="va">coreNum</span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <p></p>
<dl><dt id="arg-xlist">XList<a class="anchor" aria-label="anchor" href="#arg-xlist"></a></dt>
<dd><p>an M-length list consisting of multiple matrices with class <code>dgCMatrix</code> or <code>matrix</code> that specify the log-normalization gene expression matrix for each data sample used for PRECAST model.</p></dd>

  <dt id="arg-q">q<a class="anchor" aria-label="anchor" href="#arg-q"></a></dt>
<dd><p>a positive integer, specify the number of latent features to be extracted, default as 15.</p></dd>

  <dt id="arg-k">K<a class="anchor" aria-label="anchor" href="#arg-k"></a></dt>
<dd><p>a positive  integer allowing scalar or vector, specify the number of clusters in model fitting.</p></dd>

 <dt id="arg-adjlist">AdjList<a class="anchor" aria-label="anchor" href="#arg-adjlist"></a></dt>
<dd><p>an M-length list of sparse matrices with class <code>dgCMatrix</code>, specify the adjacency matrix used for Potts model in PRECAST. We provide this interface for those users  who would like to define the adjacency matrix by their own.</p></dd>

 <dt id="arg-adjlist-car">Adjlist_car<a class="anchor" aria-label="anchor" href="#arg-adjlist-car"></a></dt>
<dd><p>an M-length list of sparse matrices with class <code>dgCMatrix</code>, specify the adjacency matrix used for CAR model in PRECAST, default as AdjList in the Potts model. We provide this interface for those users  who would like to use the different adjacency matrix in CAR model.</p></dd>

 <dt id="arg-poslist">posList<a class="anchor" aria-label="anchor" href="#arg-poslist"></a></dt>
<dd><p>an M-length list composed by spatial coordinate matrix for each data sample.</p></dd>

 <dt id="arg-platform">platform<a class="anchor" aria-label="anchor" href="#arg-platform"></a></dt>
<dd><p>a string, specify the platform of the provided data, default as "Visium".
  There are many platforms to be supported, including ("Visuim", "ST", "SeqFISH", 'merFISH', 'slide-seqv2', 'seqscope', "HDST"). If AdjList is not given, the The platform helps to calculate the adjacency matrix by defining the neighbors.</p></dd>

  <dt id="arg-beta-grid">beta_grid<a class="anchor" aria-label="anchor" href="#arg-beta-grid"></a></dt>
<dd><p>an optional vector of positive value, the candidate set of the smoothing parameter to be searched by the grid-search optimization approach.</p></dd>

  <dt id="arg-init-nstart">init.nstart<a class="anchor" aria-label="anchor" href="#arg-init-nstart"></a></dt>
<dd><p>an optional positive integer, how many times to be run for obtaining the initial values, default as 5.</p></dd>

  <dt id="arg-maxiter-icm">maxIter_ICM<a class="anchor" aria-label="anchor" href="#arg-maxiter-icm"></a></dt>
<dd><p>an optional positive value, represents the maximum iterations of ICM.</p></dd>

  <dt id="arg-maxiter">maxIter<a class="anchor" aria-label="anchor" href="#arg-maxiter"></a></dt>
<dd><p>an optional positive value, represents the maximum iterations of EM.</p></dd>

  <dt id="arg-epsloglik">epsLogLik<a class="anchor" aria-label="anchor" href="#arg-epsloglik"></a></dt>
<dd><p>an optional positive vlaue, tolerance vlaue of relative variation rate of the observed pseudo log-loglikelihood value, defualt as '1e-5'.</p></dd>

  <dt id="arg-verbose">verbose<a class="anchor" aria-label="anchor" href="#arg-verbose"></a></dt>
<dd><p>an optional logical value, whether output the information of the ICM-EM algorithm.</p></dd>

  <dt id="arg-mix-prop-heter">mix_prop_heter<a class="anchor" aria-label="anchor" href="#arg-mix-prop-heter"></a></dt>
<dd><p>an optional logical value, specify whether betar are distict, default as <code>TRUE</code>.</p></dd>

  <dt id="arg-sigma-equal">Sigma_equal<a class="anchor" aria-label="anchor" href="#arg-sigma-equal"></a></dt>
<dd><p>an optional logical value, specify whether Sigmaks are equal, default as FALSE.</p></dd>

  <dt id="arg-sigma-diag">Sigma_diag<a class="anchor" aria-label="anchor" href="#arg-sigma-diag"></a></dt>
<dd><p>an optional logical value, specify whether Sigmaks are diagonal matrices, default as <code>TRUE</code>.</p></dd>

  <dt id="arg-error-heter">error_heter<a class="anchor" aria-label="anchor" href="#arg-error-heter"></a></dt>
<dd><p>an optional logical value, whether use the heterogenous error for DR-SC model, default as <code>TRUE</code>. If <code>error_heter=FALSE</code>, then the homogenuous error is used for probabilistic PCA model in PRECAST.</p></dd>

  <dt id="arg-sp-">Sp2<a class="anchor" aria-label="anchor" href="#arg-sp-"></a></dt>
<dd><p>an optional logical value, whether add the ICAR model component in the model, default as TRUE. We provide this interface for those users  who don't want to include the  ICAR model.</p></dd>

  <dt id="arg-wpca-int">wpca_int<a class="anchor" aria-label="anchor" href="#arg-wpca-int"></a></dt>
<dd><p>an optional logical value, means whether use the weighted PCA to obtain the initial values of loadings and other paramters, default as <code>FALSE</code> which means the ordinary PCA  is used.</p></dd>

  <dt id="arg-int-model">int.model<a class="anchor" aria-label="anchor" href="#arg-int-model"></a></dt>
<dd><p>an optional string, specify which model is used in evaluting the initial values for PRECAST; "kmeans" and "mclust" are supported, default as "kmeans".</p></dd>

  <dt id="arg-seed">seed<a class="anchor" aria-label="anchor" href="#arg-seed"></a></dt>
<dd><p>an optional integer, the random seed in fitting PRECAST model.</p></dd>

  <dt id="arg-corenum">coreNum<a class="anchor" aria-label="anchor" href="#arg-corenum"></a></dt>
<dd><p>an optional positive integer, means the number of thread used in parallel computating.</p></dd>

  <dt id="arg-corenum-int">coreNum_int<a class="anchor" aria-label="anchor" href="#arg-corenum-int"></a></dt>
<dd><p>an optional positive integer, means the number of cores used in parallel computation for initial values when <code>K</code> is a vector, default as same as <code>coreNum</code>.</p></dd>

</dl></div>
    <div id="details">
    <h2>Details</h2>
    <p>Nothing</p>
    </div>
    <div id="value">
    <h2>Value</h2>
    <p>ICM.EM returns a <a href="https://rdrr.io/r/base/list.html" class="external-link">list</a> with class "SeqKiDRSC_Object" with the number of components equal to the length of <code>K</code>, where each component includes the model fitting results for one number of cluster and is a list consisting of following components:</p>
<dl><dt>cluster</dt>
<dd><p>an M-length list that includes the inferred class labels for each data sample.</p></dd>

<dt>hZ</dt>
<dd><p>an M-length list that includes the batch corrected low-dimensional embeddings for each data sample.</p></dd>

<dt>hV</dt>
<dd><p>an M-length list that includes the estimate the ICAR component for each sample.</p></dd>

<dt>Rf</dt>
<dd><p>an M-length list that includes the posterior probability of domain clusters for each sample.</p></dd>

<dt>beta</dt>
<dd><p>an M-length vector that includes the estimated smoothing parameters for each sample.</p></dd>

<dt>Mu</dt>
<dd><p>mean vectors of mixtures components.</p></dd>

<dt>Sigma</dt>
<dd><p>covariance matrix of mixtures components.</p></dd>

<dt>W</dt>
<dd><p>estimated loading matrix</p></dd>

<dt>Lam</dt>
<dd><p>estimated variance of errors in probabilistic PCA model</p></dd>

<dt>loglik</dt>
<dd><p>pseudo observed log-likelihood.</p></dd>

</dl></div>
    <div id="references">
    <h2>References</h2>
    <p><a href="https://www.nature.com/articles/s41467-023-35947-w" class="external-link">Wei Liu, Liao, X., Luo, Z. et al, Jin Liu* (2023). Probabilistic embedding, clustering, and alignment for integrating spatial transcriptomics data with PRECAST. Nature Communications, 14, 296</a></p>
    </div>
    <div id="author">
    <h2>Author</h2>
    <p>Wei Liu</p>
    </div>
    <div id="note">
    <h2>Note</h2>
    <p>nothing</p>
    </div>
    <div id="see-also">
    <h2>See also</h2>
    <div class="dont-index"><p>None</p></div>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># \donttest{</span></span></span>
<span class="r-in"><span>  <span class="co">## we generate the spatial transcriptomics data with lattice neighborhood, i.e. ST platform.</span></span></span>
<span class="r-in"><span>  <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://Matrix.R-forge.R-project.org" class="external-link">Matrix</a></span><span class="op">)</span></span></span>
<span class="r-in"><span>  <span class="va">q</span> <span class="op">&lt;-</span> <span class="fl">10</span>; <span class="va">K</span> <span class="op">&lt;-</span> <span class="fl">4</span></span></span>
<span class="r-in"><span>  <span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">PRECASTObj</span><span class="op">)</span></span></span>
<span class="r-in"><span>  <span class="va">posList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">PRECASTObj</span><span class="op">@</span><span class="va">seulist</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="va">x</span><span class="op">$</span><span class="va">row</span>, <span class="va">x</span><span class="op">$</span><span class="va">col</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span>  <span class="va">AdjList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">posList</span>, <span class="va">getAdj_reg</span>, platform<span class="op">=</span><span class="st">'ST'</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Neighbors were identified for 900 out of 900 spots.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Neighbors were identified for 900 out of 900 spots.</span>
<span class="r-in"><span>  <span class="va">XList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">PRECASTObj</span><span class="op">@</span><span class="va">seulist</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">x</span><span class="op">[[</span><span class="st">'RNA'</span><span class="op">]</span><span class="op">]</span><span class="op">@</span><span class="va">data</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span>  <span class="va">XList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">XList</span>, <span class="va">scale</span>, scale<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span></span>
<span class="r-in"><span>  <span class="co">## For illustration, maxIter is set to 4</span></span></span>
<span class="r-in"><span>  <span class="va">resList</span> <span class="op">&lt;-</span> <span class="fu">ICM.EM</span><span class="op">(</span><span class="va">XList</span>,AdjList <span class="op">=</span> <span class="va">AdjList</span>, maxIter<span class="op">=</span><span class="fl">4</span>,</span></span>
<span class="r-in"><span>                   q<span class="op">=</span><span class="va">q</span>, K<span class="op">=</span><span class="va">K</span>, verbose<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> -----Intergrative data info.: 2 samples, 97 genes X 1800 spots------</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> -----Numbers of spots are: 900, 900-----</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Starting computing initial values using kmeans ...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2025-09-29 12:46:16.929582 : ***** Initialization finished!, 0.017 mins elapsed.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> ----Fitting PRECAST model...----------------</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> iter = 2, loglik= -60435.160156, dloglik=0.999972 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> iter = 3, loglik= -53606.203125, dloglik=0.112996 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> iter = 4, loglik= -51282.445312, dloglik=0.043349 </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2025-09-29 12:46:17.046719 : ***** PRECAST model fitting finished!, 0.002 mins elapsed.</span>
<span class="r-in"><span><span class="co"># }</span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Wei Liu, Yi Yang, Jin Liu.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

      </footer></div>






  </body></html>

